// @universal-robots/utilities-units/fesm2022/universal-robots-utilities-units.mjs
import Big from 'big.js';

var UnitEnum;
(function (UnitEnum) {
    UnitEnum["ACCELERATION"] = "ACCELERATION";
    UnitEnum["ANGULAR_ACCELERATION"] = "ANGULAR_ACCELERATION";
    UnitEnum["ANGULAR_SPEED"] = "ANGULAR_SPEED";
    UnitEnum["CURRENT"] = "CURRENT";
    UnitEnum["FORCE"] = "FORCE";
    UnitEnum["LENGTH"] = "LENGTH";
    UnitEnum["MASS"] = "MASS";
    UnitEnum["MOMENTUM"] = "MOMENTUM";
    UnitEnum["PLANE_ANGLE"] = "PLANE_ANGLE";
    UnitEnum["POWER"] = "POWER";
    UnitEnum["SPEED"] = "SPEED";
    UnitEnum["TEMPERATURE"] = "TEMPERATURE";
    UnitEnum["TIME"] = "TIME";
    UnitEnum["TORQUE"] = "TORQUE";
    UnitEnum["VOLTAGE"] = "VOLTAGE";
    UnitEnum["MOMENT_INERTIA"] = "MOMENT_INERTIA";
})(UnitEnum || (UnitEnum = {}));

/**
 * factor for 1 degree in radians.
 */
const degreeScaleFactor = Number.parseFloat(new Big(Math.PI).div(180).toFixed());
/**
 * factor for 1 kilogram meter in newton meters.
 */
const gravitationalAcceleration = 9.80665;
/**
 * Length unit type.
 */
class Length {
    constructor() {
        this.quantity = UnitEnum.LENGTH;
        this.length = true; // needed for generics type checking
    }
}
/**
 * Angle unit type.
 */
class Angle {
    constructor() {
        this.quantity = UnitEnum.PLANE_ANGLE;
        this.angle = true; // needed for generics type checking
    }
}
/**
 * Mass unit type.
 */
class Mass {
    constructor() {
        this.quantity = UnitEnum.MASS;
        this.mass = true; // needed for generics type checking
    }
}
/**
 * Voltage unit type.
 */
class Voltage {
    constructor() {
        this.quantity = UnitEnum.VOLTAGE;
        this.voltage = true;
    }
}
/**
 * Current unit type.
 */
class Current {
    constructor() {
        this.quantity = UnitEnum.CURRENT;
        this.current = true;
    }
}
/**
 * Time unit type.
 */
class Time {
    constructor() {
        this.quantity = UnitEnum.TIME;
        this.time = true;
    }
}
/**
 * Force unit type.
 */
class Force {
    constructor() {
        this.quantity = UnitEnum.FORCE;
        this.force = true;
    }
}
/**
 * Speed unit type.
 */
class Speed {
    constructor() {
        this.quantity = UnitEnum.SPEED;
        this.speed = true;
    }
}
/**
 * Angular speed unit type.
 */
class AngularSpeed {
    constructor() {
        this.quantity = UnitEnum.ANGULAR_SPEED;
        this.angularSpeed = true;
    }
}
/**
 * Acceleration unit type.
 */
class Acceleration {
    constructor() {
        this.quantity = UnitEnum.ACCELERATION;
        this.acceleration = true;
    }
}
/**
 * Angular acceleration unit type.
 */
class AngularAcceleration {
    constructor() {
        this.quantity = UnitEnum.ANGULAR_ACCELERATION;
        this.angularAcceleration = true;
    }
}
/**
 * Power unit type.
 */
class Power {
    constructor() {
        this.quantity = UnitEnum.POWER;
        this.power = true;
    }
}
/**
 * Torque unit type.
 */
class Torque {
    constructor() {
        this.quantity = UnitEnum.TORQUE;
        this.torque = true;
    }
}
/**
 * Momentum unit type.
 */
class Momentum {
    constructor() {
        this.quantity = UnitEnum.MOMENTUM;
        this.momentum = true;
    }
}
/**
 * Moment of Inertia unit type.
 */
class MomentOfInertia {
    constructor() {
        this.quantity = UnitEnum.MOMENT_INERTIA;
        this.momentOfInertia = true;
    }
}
/**
 * Temperature unit type
 */
class Temperature {
    constructor() {
        this.quantity = UnitEnum.TEMPERATURE;
        this.temperature = true;
    }
}
class DerivedUnit {
    /** Creates a new derived unit with the specified new unit type, using the specified input units to derive it
     *
     * @param quantity The quantity type of the new unit
     * @param numeratorUnit The numerator of the new unit
     * @param denominatorUnit The denominator of the new unit
     * @param significantDigits The number of significant digits for the new unit
     */
    constructor(quantity, numeratorUnit, denominatorUnit, significantDigits) {
        this.quantity = quantity;
        this.numeratorUnit = numeratorUnit;
        this.denominatorUnit = denominatorUnit;
        this.significantDigits = significantDigits;
        this.label = this.numeratorUnit.label.endsWith(denominatorUnit.label)
            ? `${this.numeratorUnit.label}^2`
            : `${this.numeratorUnit.label}/${this.denominatorUnit.label}`;
        this.symbol = this.numeratorUnit.symbol.endsWith(denominatorUnit.symbol)
            ? `${this.numeratorUnit.symbol}²`
            : `${this.numeratorUnit.symbol}/${this.denominatorUnit.symbol}`;
        this.scaleFactor = this.numeratorUnit.scaleFactor * this.denominatorUnit.scaleFactor;
    }
}
// Base units
/**
 * Unit of length in mm.
 */
const mm = {
    quantity: new Length(),
    scaleFactor: 0.001,
    significantDigits: 1,
    label: 'mm',
    symbol: 'mm',
};
/**
 * Unit of length in m.
 */
const m = {
    quantity: new Length(),
    scaleFactor: 1,
    significantDigits: 4,
    label: 'm',
    symbol: 'm',
};
/**
 * Array of all supported length units.
 */
const lengthUnits = [mm, m];
/**
 * Unit of length in radians.
 */
const rad = {
    quantity: new Angle(),
    scaleFactor: 1,
    significantDigits: 4,
    label: 'rad',
    symbol: 'rad',
};
/**
 * Unit of length in degrees.
 */
const deg = {
    quantity: new Angle(),
    scaleFactor: degreeScaleFactor,
    significantDigits: 2,
    label: 'deg',
    symbol: '°',
};
/**
 * Array of all supported angle units.
 */
const angleUnits = [rad, deg];
/**
 * Unit of mass in grams.
 */
const g = {
    quantity: new Mass(),
    scaleFactor: 0.001,
    significantDigits: 0,
    label: 'g',
    symbol: 'g',
};
/**
 * Unit of mass in kilograms.
 */
const kg = {
    quantity: new Mass(),
    scaleFactor: 1,
    significantDigits: 3,
    label: 'kg',
    symbol: 'kg',
};
/**
 * Array of all supported mass units.
 */
const massUnits = [g, kg];
/**
 * Unit of voltage in volts.
 */
const V = {
    quantity: new Voltage(),
    scaleFactor: 1,
    significantDigits: 3,
    label: 'V',
    symbol: 'V',
};
/**
 * Unit of voltage in millivolts.
 */
const mV = {
    quantity: new Voltage(),
    scaleFactor: 0.001,
    significantDigits: 0,
    label: 'mV',
    symbol: 'mV',
};
/**
 * Array of all supported voltage units.
 */
const voltageUnits = [V, mV];
/**
 * Unit of current in ampere.
 */
const A = {
    quantity: new Current(),
    scaleFactor: 1,
    significantDigits: 5,
    label: 'A',
    symbol: 'A',
};
/**
 * Unit of current in milliampere.
 */
const mA = {
    quantity: new Current(),
    scaleFactor: 0.001,
    significantDigits: 2,
    label: 'mA',
    symbol: 'mA',
};
/**
 * Array of all supported current units.
 */
const currentUnits = [A, mA];
/**
 * Unit of time in seconds.
 */
const s = {
    quantity: new Time(),
    scaleFactor: 1,
    significantDigits: 2,
    label: 's',
    symbol: 's',
};
/**
 * Unit of time in milliseconds.
 */
const ms = {
    quantity: new Time(),
    scaleFactor: 0.001,
    significantDigits: 0,
    label: 'ms',
    symbol: 'ms',
};
/**
 * Array of all supported time units.
 */
const timeUnits = [s, ms];
/**
 * Unit of force in newton.
 */
const N = {
    quantity: new Force(),
    scaleFactor: 1,
    significantDigits: 1,
    label: 'N',
    symbol: 'N',
};
/**
 * Array of all supported force units.
 */
const forceUnits = [N];
/**
 * Unit of power in watts.
 */
const W = {
    quantity: new Power(),
    scaleFactor: 1,
    significantDigits: 1,
    label: 'W',
    symbol: 'W',
};
/**
 * Array of all supported power units.
 */
const powerUnits = [W];
/**
 * Unit of torque in kilogram meters.
 */
const kgM = {
    quantity: new Torque(),
    scaleFactor: gravitationalAcceleration,
    significantDigits: 1,
    label: 'kg*m',
    symbol: 'kg·m',
};
/**
 * Unit of moment of inertia in kilogram meters squared
 */
const kgMSq = {
    quantity: new MomentOfInertia(),
    scaleFactor: 1,
    significantDigits: 6,
    label: 'kg*m*m',
    symbol: 'kg·m²',
};
/**
 * Unit of temperature in degrees Celsius
 */
const degreesCelsius = {
    quantity: new Temperature(),
    scaleFactor: 1,
    significantDigits: 2,
    label: 'C',
    symbol: '°',
};
/**
 * Array of all supported torque units.
 */
const torqueUnits = [kgM];
// Derived Units
/**
 * Unit of speed in meters per second.
 */
const mPerSec = new DerivedUnit(new Speed(), m, s, 5);
/**
 * Unit of speed in millimeters per second.
 */
const mmPerSec = new DerivedUnit(new Speed(), mm, s, 2);
/**
 * Array of all supported speed units.
 */
const speedUnits = [mPerSec, mmPerSec];
/**
 * Unit of angular speed in radiant per second.
 */
const radPerSec = new DerivedUnit(new AngularSpeed(), rad, s, 3);
/**
 * Unit of angular speed in degree per second.
 */
const degPerSec = new DerivedUnit(new AngularSpeed(), deg, s, 2);
/**
 * Array of all supported angular speed units.
 */
const angularSpeedUnits = [radPerSec, degPerSec];
/**
 * Unit of acceleration in meter per second squared.
 */
const mPerSecSq = new DerivedUnit(new Acceleration(), mPerSec, s, 5);
/**
 * Unit of acceleration in millimeter per second squared.
 */
const mmPerSecSq = new DerivedUnit(new Acceleration(), mmPerSec, s, 2);
/**
 * Array of all supported acceleration units.
 */
const accelerationUnits = [mPerSecSq, mmPerSecSq];
/**
 * Unit of angular acceleration in radian per second squared.
 */
const radPerSecSq = new DerivedUnit(new AngularAcceleration(), radPerSec, s, 3);
/**
 * Unit of angular acceleration in degree per second squared.
 */
const degPerSecSq = new DerivedUnit(new AngularAcceleration(), degPerSec, s, 2);
/**
 * Array of all supported angular acceleration units.
 */
const angularAccelerationUnits = [radPerSecSq, degPerSecSq];
/**
 * Unit of momentum in kilogram meters per second.
 */
const kgMPerSec = new DerivedUnit(new Momentum(), kgM, s, 1);
/**
 * Array of all supported momentum units.
 */
const momentumUnits = [kgMPerSec];
const momentInertiaUnits = [kgMSq];
/**
 * Array of all supported temperature units
 */
const temperatureUnits = [degreesCelsius];
/**
 * All units.
 */
const allUnits = [
    ...lengthUnits,
    ...angleUnits,
    ...massUnits,
    ...voltageUnits,
    ...currentUnits,
    ...speedUnits,
    ...angularSpeedUnits,
    ...accelerationUnits,
    ...angularAccelerationUnits,
    ...timeUnits,
    ...torqueUnits,
    ...forceUnits,
    ...powerUnits,
    ...momentumUnits,
    ...momentInertiaUnits,
    ...temperatureUnits,
];
/**
 * Map from unit enumeration to SI unit.
 */
const CONCEPT_UNITS = {
    [UnitEnum.LENGTH]: m,
    [UnitEnum.MASS]: kg,
    [UnitEnum.PLANE_ANGLE]: rad,
    [UnitEnum.VOLTAGE]: V,
    [UnitEnum.CURRENT]: A,
    [UnitEnum.ACCELERATION]: mPerSecSq,
    [UnitEnum.ANGULAR_ACCELERATION]: radPerSecSq,
    [UnitEnum.SPEED]: mPerSec,
    [UnitEnum.ANGULAR_SPEED]: radPerSec,
    [UnitEnum.TIME]: s,
    [UnitEnum.TORQUE]: kgM,
    [UnitEnum.FORCE]: N,
    [UnitEnum.POWER]: W,
    [UnitEnum.MOMENTUM]: kgMPerSec,
    [UnitEnum.MOMENT_INERTIA]: kgMSq,
    [UnitEnum.TEMPERATURE]: degreesCelsius,
};

/**
 * FixedPointNumber.
 *
 * Provides representations of a value.
 */
class FixedPointNumber {
    /**
     *
     * @param value. The actual value.
     * @param unit. The unit for the provided value.
     */
    constructor(value, unit) {
        this.bigValue = new Big(value);
        const fixed = this.bigValue.toFixed(unit.significantDigits);
        const decimalPoint = unit.significantDigits === 0 ? 0 : 1;
        this.representation = fixed.match(`-0[.]{${decimalPoint}}0{${unit.significantDigits}}$`) ? fixed.substring(1) : fixed;
        this.unit = unit;
        this.value = Number.parseFloat(this.bigValue.toFixed());
    }
    /**
     * return value in number format.
     */
    getValue() {
        return this.value;
    }
    /**
     * return value in big value format.
     */
    getBigValue() {
        return this.bigValue;
    }
    /**
     * return formatted string representation of the value.
     */
    getDisplayValue() {
        return this.representation;
    }
    /**
     * return the values unit.
     */
    getUnit() {
        return this.unit;
    }
}

/**
 * Quantity.
 *
 * For a specific unit type, converts between units.
 */
class Quantity {
    /**
     *
     * @param value to convert.
     * @param unit for the provided value.
     */
    constructor(value, unit) {
        this.value = value.mul(unit.scaleFactor);
    }
    /**
     *
     * @param unit. the value should be converted to this unit.
     */
    getAs(unit) {
        return this.value.div(unit.scaleFactor);
    }
}

/**
 *
 * @param value. The value to convert from
 * @param to. The unit to convert to.
 */
function converter(value, to) {
    const toUnit = value.getUnit();
    const quantity = new Quantity(value.getBigValue(), toUnit);
    return new FixedPointNumber(quantity.getAs(to), to);
}
/**
 * Converter for mass unit types.
 */
const massConverter = converter;
/**
 * Converter for length unit types.
 */
const lengthConverter = converter;
/**
 * Converter for angle unit types.
 */
const angleConverter = converter;
/**
 * Converter for angular speed unit types.
 */
const angularSpeedConverter = converter;
/**
 * Converter for voltage unit types.
 */
const voltageConverter = converter;
/**
 * Converter for current unit types.
 */
const currentConverter = converter;
/**
 * Converter for force unit types.
 */
const forceConverter = converter;
/**
 * Converter for power unit types.
 */
const powerConverter = converter;
/**
 * Converter for torque unit types.
 */
const torqueConverter = converter;
/**
 * Converter for momentum unit types.
 */
const momentumConverter = converter;
/**
 * Converter for speed unit types.
 */
const speedConverter = converter;
/**
 * Converter for time unit types.
 */
const timeConverter = converter;

/*
 * Public API Surface of utilities-units
 */

/**
 * Generated bundle index. Do not edit.
 */

export { A, Acceleration, Angle, AngularAcceleration, AngularSpeed, CONCEPT_UNITS, Current, DerivedUnit, FixedPointNumber, Force, Length, Mass, MomentOfInertia, Momentum, N, Power, Quantity, Speed, Temperature, Time, Torque, UnitEnum, V, Voltage, W, accelerationUnits, allUnits, angleConverter, angleUnits, angularAccelerationUnits, angularSpeedConverter, angularSpeedUnits, converter, currentConverter, currentUnits, deg, degPerSec, degPerSecSq, degreeScaleFactor, degreesCelsius, forceConverter, forceUnits, g, gravitationalAcceleration, kg, kgM, kgMPerSec, kgMSq, lengthConverter, lengthUnits, m, mA, mPerSec, mPerSecSq, mV, massConverter, massUnits, mm, mmPerSec, mmPerSecSq, momentInertiaUnits, momentumConverter, momentumUnits, ms, powerConverter, powerUnits, rad, radPerSec, radPerSecSq, s, speedConverter, speedUnits, temperatureUnits, timeConverter, timeUnits, torqueConverter, torqueUnits, voltageConverter, voltageUnits };
