// @universal-robots/ros2-interfaces/index.ts
/* eslint-disable */
// These files were generated using "ros-typescript-generator"
export interface IRosTypeGeometryMsgsAccel {
  linear: IRosTypeGeometryMsgsVector3;
  angular: IRosTypeGeometryMsgsVector3;
}

export interface IRosTypeGeometryMsgsAccelStamped {
  header: IRosTypeStdMsgsHeader;
  accel: IRosTypeGeometryMsgsAccel;
}

export interface IRosTypeGeometryMsgsAccelWithCovariance {
  accel: IRosTypeGeometryMsgsAccel;
  covariance: number[];
}

export interface IRosTypeGeometryMsgsAccelWithCovarianceStamped {
  header: IRosTypeStdMsgsHeader;
  accel: IRosTypeGeometryMsgsAccelWithCovariance;
}

export interface IRosTypeGeometryMsgsInertia {
  m: number;
  com: IRosTypeGeometryMsgsVector3;
  ixx: number;
  ixy: number;
  ixz: number;
  iyy: number;
  iyz: number;
  izz: number;
}

export interface IRosTypeGeometryMsgsInertiaStamped {
  header: IRosTypeStdMsgsHeader;
  inertia: IRosTypeGeometryMsgsInertia;
}

export interface IRosTypeGeometryMsgsPoint {
  x: number;
  y: number;
  z: number;
}

export interface IRosTypeGeometryMsgsPoint32 {
  x: number;
  y: number;
  z: number;
}

export interface IRosTypeGeometryMsgsPointStamped {
  header: IRosTypeStdMsgsHeader;
  point: IRosTypeGeometryMsgsPoint;
}

export interface IRosTypeGeometryMsgsPolygon {
  points: IRosTypeGeometryMsgsPoint32[];
}

export interface IRosTypeGeometryMsgsPolygonStamped {
  header: IRosTypeStdMsgsHeader;
  polygon: IRosTypeGeometryMsgsPolygon;
}

export interface IRosTypeGeometryMsgsPose {
  position: IRosTypeGeometryMsgsPoint;
  orientation: IRosTypeGeometryMsgsQuaternion;
}

export interface IRosTypeGeometryMsgsPose2D {
  x: number;
  y: number;
  theta: number;
}

export interface IRosTypeGeometryMsgsPoseArray {
  header: IRosTypeStdMsgsHeader;
  poses: IRosTypeGeometryMsgsPose[];
}

export interface IRosTypeGeometryMsgsPoseStamped {
  header: IRosTypeStdMsgsHeader;
  pose: IRosTypeGeometryMsgsPose;
}

export interface IRosTypeGeometryMsgsPoseWithCovariance {
  pose: IRosTypeGeometryMsgsPose;
  covariance: number[];
}

export interface IRosTypeGeometryMsgsPoseWithCovarianceStamped {
  header: IRosTypeStdMsgsHeader;
  pose: IRosTypeGeometryMsgsPoseWithCovariance;
}

export interface IRosTypeGeometryMsgsQuaternion {
  x: number;
  y: number;
  z: number;
  w: number;
}

export interface IRosTypeGeometryMsgsQuaternionStamped {
  header: IRosTypeStdMsgsHeader;
  quaternion: IRosTypeGeometryMsgsQuaternion;
}

export interface IRosTypeGeometryMsgsTransform {
  translation: IRosTypeGeometryMsgsVector3;
  rotation: IRosTypeGeometryMsgsQuaternion;
}

export interface IRosTypeGeometryMsgsTransformStamped {
  header: IRosTypeStdMsgsHeader;
  child_frame_id: string;
  transform: IRosTypeGeometryMsgsTransform;
}

export interface IRosTypeGeometryMsgsTwist {
  linear: IRosTypeGeometryMsgsVector3;
  angular: IRosTypeGeometryMsgsVector3;
}

export interface IRosTypeGeometryMsgsTwistStamped {
  header: IRosTypeStdMsgsHeader;
  twist: IRosTypeGeometryMsgsTwist;
}

export interface IRosTypeGeometryMsgsTwistWithCovariance {
  twist: IRosTypeGeometryMsgsTwist;
  covariance: number[];
}

export interface IRosTypeGeometryMsgsTwistWithCovarianceStamped {
  header: IRosTypeStdMsgsHeader;
  twist: IRosTypeGeometryMsgsTwistWithCovariance;
}

export interface IRosTypeGeometryMsgsVector3 {
  x: number;
  y: number;
  z: number;
}

export interface IRosTypeGeometryMsgsVector3Stamped {
  header: IRosTypeStdMsgsHeader;
  vector: IRosTypeGeometryMsgsVector3;
}

export interface IRosTypeGeometryMsgsWrench {
  force: IRosTypeGeometryMsgsVector3;
  torque: IRosTypeGeometryMsgsVector3;
}

export interface IRosTypeGeometryMsgsWrenchStamped {
  header: IRosTypeStdMsgsHeader;
  wrench: IRosTypeGeometryMsgsWrench;
}

export enum IRosTypeRclInterfacesParameterTypeConst {
  PARAMETER_NOT_SET = 0,
  PARAMETER_BOOL = 1,
  PARAMETER_INTEGER = 2,
  PARAMETER_DOUBLE = 3,
  PARAMETER_STRING = 4,
  PARAMETER_BYTE_ARRAY = 5,
  PARAMETER_BOOL_ARRAY = 6,
  PARAMETER_INTEGER_ARRAY = 7,
  PARAMETER_DOUBLE_ARRAY = 8,
  PARAMETER_STRING_ARRAY = 9,
}

export interface IRosTypeRclInterfacesDescribeParameters {
  request: IRosTypeRclInterfacesDescribeParametersRequest;
  response: IRosTypeRclInterfacesDescribeParametersResponse;
}

export interface IRosTypeRclInterfacesDescribeParametersRequest {
  names: string[];
}

export interface IRosTypeRclInterfacesDescribeParametersResponse {
  descriptors: IRosTypeRclInterfacesParameterDescriptor[];
}

export interface IRosTypeRclInterfacesFloatingPointRange {
  from_value: number;
  to_value: number;
  step: number;
}

export interface IRosTypeRclInterfacesGetParameterTypes {
  request: IRosTypeRclInterfacesGetParameterTypesRequest;
  response: IRosTypeRclInterfacesGetParameterTypesResponse;
}

export interface IRosTypeRclInterfacesGetParameterTypesRequest {
  names: string[];
}

export interface IRosTypeRclInterfacesGetParameterTypesResponse {
  types: number[];
}

export interface IRosTypeRclInterfacesGetParameters {
  request: IRosTypeRclInterfacesGetParametersRequest;
  response: IRosTypeRclInterfacesGetParametersResponse;
}

export interface IRosTypeRclInterfacesGetParametersRequest {
  names: string[];
}

export interface IRosTypeRclInterfacesGetParametersResponse {
  values: IRosTypeRclInterfacesParameterValue[];
}

export interface IRosTypeRclInterfacesIntegerRange {
  from_value: number;
  to_value: number;
  step: number;
}

export interface IRosTypeRclInterfacesListParameters {
  request: IRosTypeRclInterfacesListParametersRequest;
  response: IRosTypeRclInterfacesListParametersResponse;
}

export interface IRosTypeRclInterfacesListParametersRequest {
  prefixes: string[];
  depth: number;
}

export enum IRosTypeRclInterfacesListParametersRequestConst {
  DEPTH_RECURSIVE = 0,
}

export interface IRosTypeRclInterfacesListParametersResponse {
  result: IRosTypeRclInterfacesListParametersResult;
}

export interface IRosTypeRclInterfacesListParametersResult {
  names: string[];
  prefixes: string[];
}

export interface IRosTypeRclInterfacesLog {
  stamp: { sec: number, nanosec: number };
  level: number;
  name: string;
  msg: string;
  file: string;
  function: string;
  line: number;
}

export enum IRosTypeRclInterfacesLogConst {
  DEBUG = 10,
  INFO = 20,
  WARN = 30,
  ERROR = 40,
  FATAL = 50,
}

export interface IRosTypeRclInterfacesParameter {
  name: string;
  value: IRosTypeRclInterfacesParameterValue;
}

export interface IRosTypeRclInterfacesParameterDescriptor {
  name: string;
  type: number;
  description: string;
  additional_constraints: string;
  read_only: boolean;
  dynamic_typing: boolean;
  floating_point_range: IRosTypeRclInterfacesFloatingPointRange[];
  integer_range: IRosTypeRclInterfacesIntegerRange[];
}

export interface IRosTypeRclInterfacesParameterEvent {
  stamp: { sec: number, nanosec: number };
  node: string;
  new_parameters: IRosTypeRclInterfacesParameter[];
  changed_parameters: IRosTypeRclInterfacesParameter[];
  deleted_parameters: IRosTypeRclInterfacesParameter[];
}

export interface IRosTypeRclInterfacesParameterEventDescriptors {
  new_parameters: IRosTypeRclInterfacesParameterDescriptor[];
  changed_parameters: IRosTypeRclInterfacesParameterDescriptor[];
  deleted_parameters: IRosTypeRclInterfacesParameterDescriptor[];
}

export interface IRosTypeRclInterfacesParameterValue {
  type: number;
  bool_value: boolean;
  integer_value: number;
  double_value: number;
  string_value: string;
  byte_array_value: number[];
  bool_array_value: boolean[];
  integer_array_value: number[];
  double_array_value: number[];
  string_array_value: string[];
}

export interface IRosTypeRclInterfacesSetParameters {
  request: IRosTypeRclInterfacesSetParametersRequest;
  response: IRosTypeRclInterfacesSetParametersResponse;
}

export interface IRosTypeRclInterfacesSetParametersAtomically {
  request: IRosTypeRclInterfacesSetParametersAtomicallyRequest;
  response: IRosTypeRclInterfacesSetParametersAtomicallyResponse;
}

export interface IRosTypeRclInterfacesSetParametersAtomicallyRequest {
  parameters: IRosTypeRclInterfacesParameter[];
}

export interface IRosTypeRclInterfacesSetParametersAtomicallyResponse {
  result: IRosTypeRclInterfacesSetParametersResult;
}

export interface IRosTypeRclInterfacesSetParametersRequest {
  parameters: IRosTypeRclInterfacesParameter[];
}

export interface IRosTypeRclInterfacesSetParametersResponse {
  results: IRosTypeRclInterfacesSetParametersResult[];
}

export interface IRosTypeRclInterfacesSetParametersResult {
  successful: boolean;
  reason: string;
}

export interface IRosTypeSensorMsgsBatteryState {
  header: IRosTypeStdMsgsHeader;
  voltage: number;
  temperature: number;
  current: number;
  charge: number;
  capacity: number;
  design_capacity: number;
  percentage: number;
  power_supply_status: number;
  power_supply_health: number;
  power_supply_technology: number;
  present: boolean;
  cell_voltage: number[];
  cell_temperature: number[];
  location: string;
  serial_number: string;
}

export enum IRosTypeSensorMsgsBatteryStateConst {
  POWER_SUPPLY_STATUS_UNKNOWN = 0,
  POWER_SUPPLY_STATUS_CHARGING = 1,
  POWER_SUPPLY_STATUS_DISCHARGING = 2,
  POWER_SUPPLY_STATUS_NOT_CHARGING = 3,
  POWER_SUPPLY_STATUS_FULL = 4,
  POWER_SUPPLY_HEALTH_UNKNOWN = 0,
  POWER_SUPPLY_HEALTH_GOOD = 1,
  POWER_SUPPLY_HEALTH_OVERHEAT = 2,
  POWER_SUPPLY_HEALTH_DEAD = 3,
  POWER_SUPPLY_HEALTH_OVERVOLTAGE = 4,
  POWER_SUPPLY_HEALTH_UNSPEC_FAILURE = 5,
  POWER_SUPPLY_HEALTH_COLD = 6,
  POWER_SUPPLY_HEALTH_WATCHDOG_TIMER_EXPIRE = 7,
  POWER_SUPPLY_HEALTH_SAFETY_TIMER_EXPIRE = 8,
  POWER_SUPPLY_TECHNOLOGY_UNKNOWN = 0,
  POWER_SUPPLY_TECHNOLOGY_NIMH = 1,
  POWER_SUPPLY_TECHNOLOGY_LION = 2,
  POWER_SUPPLY_TECHNOLOGY_LIPO = 3,
  POWER_SUPPLY_TECHNOLOGY_LIFE = 4,
  POWER_SUPPLY_TECHNOLOGY_NICD = 5,
  POWER_SUPPLY_TECHNOLOGY_LIMN = 6,
}

export interface IRosTypeSensorMsgsCameraInfo {
  header: IRosTypeStdMsgsHeader;
  height: number;
  width: number;
  distortion_model: string;
  d: number[];
  k: number[];
  r: number[];
  p: number[];
  binning_x: number;
  binning_y: number;
  roi: IRosTypeSensorMsgsRegionOfInterest;
}

export interface IRosTypeSensorMsgsChannelFloat32 {
  name: string;
  values: number[];
}

export interface IRosTypeSensorMsgsCompressedImage {
  header: IRosTypeStdMsgsHeader;
  format: string;
  data: number[];
}

export interface IRosTypeSensorMsgsFluidPressure {
  header: IRosTypeStdMsgsHeader;
  fluid_pressure: number;
  variance: number;
}

export interface IRosTypeSensorMsgsIlluminance {
  header: IRosTypeStdMsgsHeader;
  illuminance: number;
  variance: number;
}

export interface IRosTypeSensorMsgsImage {
  header: IRosTypeStdMsgsHeader;
  height: number;
  width: number;
  encoding: string;
  is_bigendian: number;
  step: number;
  data: number[];
}

export interface IRosTypeSensorMsgsImu {
  header: IRosTypeStdMsgsHeader;
  orientation: IRosTypeGeometryMsgsQuaternion;
  orientation_covariance: number[];
  angular_velocity: IRosTypeGeometryMsgsVector3;
  angular_velocity_covariance: number[];
  linear_acceleration: IRosTypeGeometryMsgsVector3;
  linear_acceleration_covariance: number[];
}

export interface IRosTypeSensorMsgsJointState {
  header: IRosTypeStdMsgsHeader;
  name: string[];
  position: number[];
  velocity: number[];
  effort: number[];
}

export interface IRosTypeSensorMsgsJoy {
  header: IRosTypeStdMsgsHeader;
  axes: number[];
  buttons: number[];
}

export interface IRosTypeSensorMsgsJoyFeedback {
  type: number;
  id: number;
  intensity: number;
}

export enum IRosTypeSensorMsgsJoyFeedbackConst {
  TYPE_LED = 0,
  TYPE_RUMBLE = 1,
  TYPE_BUZZER = 2,
}

export interface IRosTypeSensorMsgsJoyFeedbackArray {
  array: IRosTypeSensorMsgsJoyFeedback[];
}

export interface IRosTypeSensorMsgsLaserEcho {
  echoes: number[];
}

export interface IRosTypeSensorMsgsLaserScan {
  header: IRosTypeStdMsgsHeader;
  angle_min: number;
  angle_max: number;
  angle_increment: number;
  time_increment: number;
  scan_time: number;
  range_min: number;
  range_max: number;
  ranges: number[];
  intensities: number[];
}

export interface IRosTypeSensorMsgsMagneticField {
  header: IRosTypeStdMsgsHeader;
  magnetic_field: IRosTypeGeometryMsgsVector3;
  magnetic_field_covariance: number[];
}

export interface IRosTypeSensorMsgsMultiDofJointState {
  header: IRosTypeStdMsgsHeader;
  joint_names: string[];
  transforms: IRosTypeGeometryMsgsTransform[];
  twist: IRosTypeGeometryMsgsTwist[];
  wrench: IRosTypeGeometryMsgsWrench[];
}

export interface IRosTypeSensorMsgsMultiEchoLaserScan {
  header: IRosTypeStdMsgsHeader;
  angle_min: number;
  angle_max: number;
  angle_increment: number;
  time_increment: number;
  scan_time: number;
  range_min: number;
  range_max: number;
  ranges: IRosTypeSensorMsgsLaserEcho[];
  intensities: IRosTypeSensorMsgsLaserEcho[];
}

export interface IRosTypeSensorMsgsNavSatFix {
  header: IRosTypeStdMsgsHeader;
  status: IRosTypeSensorMsgsNavSatStatus;
  latitude: number;
  longitude: number;
  altitude: number;
  position_covariance: number[];
  position_covariance_type: number;
}

export enum IRosTypeSensorMsgsNavSatFixConst {
  COVARIANCE_TYPE_UNKNOWN = 0,
  COVARIANCE_TYPE_APPROXIMATED = 1,
  COVARIANCE_TYPE_DIAGONAL_KNOWN = 2,
  COVARIANCE_TYPE_KNOWN = 3,
}

export interface IRosTypeSensorMsgsNavSatStatus {
  status: number;
  service: number;
}

export enum IRosTypeSensorMsgsNavSatStatusConst {
  STATUS_NO_FIX = -1,
  STATUS_FIX = 0,
  STATUS_SBAS_FIX = 1,
  STATUS_GBAS_FIX = 2,
  SERVICE_GPS = 1,
  SERVICE_GLONASS = 2,
  SERVICE_COMPASS = 4,
  SERVICE_GALILEO = 8,
}

export interface IRosTypeSensorMsgsPointCloud {
  header: IRosTypeStdMsgsHeader;
  points: IRosTypeGeometryMsgsPoint32[];
  channels: IRosTypeSensorMsgsChannelFloat32[];
}

export interface IRosTypeSensorMsgsPointCloud2 {
  header: IRosTypeStdMsgsHeader;
  height: number;
  width: number;
  fields: IRosTypeSensorMsgsPointField[];
  is_bigendian: boolean;
  point_step: number;
  row_step: number;
  data: number[];
  is_dense: boolean;
}

export interface IRosTypeSensorMsgsPointField {
  name: string;
  offset: number;
  datatype: number;
  count: number;
}

export enum IRosTypeSensorMsgsPointFieldConst {
  INT8 = 1,
  UINT8 = 2,
  INT16 = 3,
  UINT16 = 4,
  INT32 = 5,
  UINT32 = 6,
  FLOAT32 = 7,
  FLOAT64 = 8,
}

export interface IRosTypeSensorMsgsRange {
  header: IRosTypeStdMsgsHeader;
  radiation_type: number;
  field_of_view: number;
  min_range: number;
  max_range: number;
  range: number;
}

export enum IRosTypeSensorMsgsRangeConst {
  ULTRASOUND = 0,
  INFRARED = 1,
}

export interface IRosTypeSensorMsgsRegionOfInterest {
  x_offset: number;
  y_offset: number;
  height: number;
  width: number;
  do_rectify: boolean;
}

export interface IRosTypeSensorMsgsRelativeHumidity {
  header: IRosTypeStdMsgsHeader;
  relative_humidity: number;
  variance: number;
}

export interface IRosTypeSensorMsgsSetCameraInfo {
  request: IRosTypeSensorMsgsSetCameraInfoRequest;
  response: IRosTypeSensorMsgsSetCameraInfoResponse;
}

export interface IRosTypeSensorMsgsSetCameraInfoRequest {
  camera_info: IRosTypeSensorMsgsCameraInfo;
}

export interface IRosTypeSensorMsgsSetCameraInfoResponse {
  success: boolean;
  status_message: string;
}

export interface IRosTypeSensorMsgsTemperature {
  header: IRosTypeStdMsgsHeader;
  temperature: number;
  variance: number;
}

export interface IRosTypeSensorMsgsTimeReference {
  header: IRosTypeStdMsgsHeader;
  time_ref: { sec: number, nanosec: number };
  source: string;
}

export interface IRosTypeStdMsgsBool {
  data: boolean;
}

export interface IRosTypeStdMsgsByte {
  data: number;
}

export interface IRosTypeStdMsgsByteMultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdMsgsChar {
  data: number;
}

export interface IRosTypeStdMsgsColorRgba {
  r: number;
  g: number;
  b: number;
  a: number;
}

export interface IRosTypeStdMsgsFloat32 {
  data: number;
}

export interface IRosTypeStdMsgsFloat32MultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdMsgsFloat64 {
  data: number;
}

export interface IRosTypeStdMsgsFloat64MultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdMsgsHeader {
  stamp: { sec: number, nanosec: number };
  frame_id: string;
}

export interface IRosTypeStdMsgsInt16 {
  data: number;
}

export interface IRosTypeStdMsgsInt16MultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdMsgsInt32 {
  data: number;
}

export interface IRosTypeStdMsgsInt32MultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdMsgsInt64 {
  data: number;
}

export interface IRosTypeStdMsgsInt64MultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdMsgsInt8 {
  data: number;
}

export interface IRosTypeStdMsgsInt8MultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdMsgsMultiArrayDimension {
  label: string;
  size: number;
  stride: number;
}

export interface IRosTypeStdMsgsMultiArrayLayout {
  dim: IRosTypeStdMsgsMultiArrayDimension[];
  data_offset: number;
}

export interface IRosTypeStdMsgsString {
  data: string;
}

export interface IRosTypeStdMsgsUInt16 {
  data: number;
}

export interface IRosTypeStdMsgsUInt16MultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdMsgsUInt32 {
  data: number;
}

export interface IRosTypeStdMsgsUInt32MultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdMsgsUInt64 {
  data: number;
}

export interface IRosTypeStdMsgsUInt64MultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdMsgsUInt8 {
  data: number;
}

export interface IRosTypeStdMsgsUInt8MultiArray {
  layout: IRosTypeStdMsgsMultiArrayLayout;
  data: number[];
}

export interface IRosTypeStdSrvsSetBool {
  request: IRosTypeStdSrvsSetBoolRequest;
  response: IRosTypeStdSrvsSetBoolResponse;
}

export interface IRosTypeStdSrvsSetBoolRequest {
  data: boolean;
}

export interface IRosTypeStdSrvsSetBoolResponse {
  success: boolean;
  message: string;
}

export interface IRosTypeStdSrvsTrigger {
  response: IRosTypeStdSrvsTriggerResponse;
}

export interface IRosTypeStdSrvsTriggerResponse {
  success: boolean;
  message: string;
}

export interface IRosTypeTf2MsgsTf2Error {
  error: number;
  error_string: string;
}

export enum IRosTypeTf2MsgsTf2ErrorConst {
  NO_ERROR = 0,
  LOOKUP_ERROR = 1,
  CONNECTIVITY_ERROR = 2,
  EXTRAPOLATION_ERROR = 3,
  INVALID_ARGUMENT_ERROR = 4,
  TIMEOUT_ERROR = 5,
  TRANSFORM_ERROR = 6,
}

export interface IRosTypeTf2MsgsTfMessage {
  transforms: IRosTypeGeometryMsgsTransformStamped[];
}

export interface IRosTypeUrinterfacesActiveTcp {
  header: IRosTypeStdMsgsHeader;
  data: string;
}

export interface IRosTypeUrinterfacesAnalog {
  pin: number;
  domain: number;
  value: number;
}

export enum IRosTypeUrinterfacesAnalogConst {
  CURRENT = 0,
  VOLTAGE = 1,
}

export interface IRosTypeUrinterfacesAnalogIo {
  analog_inputs: IRosTypeUrinterfacesAnalog[];
  analog_outputs: IRosTypeUrinterfacesAnalog[];
}

export interface IRosTypeUrinterfacesAnalogIoStateStamped {
  header: IRosTypeStdMsgsHeader;
  io_state: IRosTypeUrinterfacesAnalogIo;
}

export interface IRosTypeUrinterfacesAnalogStamped {
  header: IRosTypeStdMsgsHeader;
  data: IRosTypeUrinterfacesAnalog;
}

export interface IRosTypeUrinterfacesBoolStamped {
  header: IRosTypeStdMsgsHeader;
  data: boolean;
}

export interface IRosTypeUrinterfacesControlCycle {
  header: IRosTypeStdMsgsHeader;
  data: number;
}

export interface IRosTypeUrinterfacesCurrent {
  header: IRosTypeStdMsgsHeader;
  data: number;
}

export interface IRosTypeUrinterfacesDigital {
  pin: number;
  state: boolean;
}

export interface IRosTypeUrinterfacesDigitalIo {
  digital_inputs: IRosTypeUrinterfacesDigital[];
  digital_outputs: IRosTypeUrinterfacesDigital[];
}

export interface IRosTypeUrinterfacesDigitalIoStateStamped {
  header: IRosTypeStdMsgsHeader;
  configurable_io: IRosTypeUrinterfacesDigitalIo;
  standard_io: IRosTypeUrinterfacesDigitalIo;
}

export interface IRosTypeUrinterfacesDigitalStamped {
  header: IRosTypeStdMsgsHeader;
  data: IRosTypeUrinterfacesDigital;
}

export interface IRosTypeUrinterfacesDoubleMultiArrayStamped {
  header: IRosTypeStdMsgsHeader;
  data: number[];
}

export interface IRosTypeUrinterfacesFloat32Stamped {
  header: IRosTypeStdMsgsHeader;
  data: number;
}

export interface IRosTypeUrinterfacesFloat64Stamped {
  header: IRosTypeStdMsgsHeader;
  data: number;
}

export interface IRosTypeUrinterfacesJointMode {
  header: IRosTypeStdMsgsHeader;
  joint_modes: number[];
  tool_mode: number;
}

export enum IRosTypeUrinterfacesJointModeConst {
  JOINT_MODE_RESET = 235,
  JOINT_MODE_SHUTTING_DOWN = 236,
  JOINT_MODE_PART_D_CALIBRATION = 237,
  JOINT_MODE_BACKDRIVE = 238,
  JOINT_MODE_POWER_OFF = 239,
  JOINT_MODE_READY_FOR_POWER_OFF = 240,
  JOINT_MODE_NOT_RESPONDING = 245,
  JOINT_MODE_MOTOR_INITIALISATION = 246,
  JOINT_MODE_BOOTING = 247,
  JOINT_MODE_PART_D_CALIBRATION_ERROR = 248,
  JOINT_MODE_BOOTLOADER = 249,
  JOINT_MODE_CALIBRATION = 250,
  JOINT_MODE_VIOLATION = 251,
  JOINT_MODE_FAULT = 252,
  JOINT_MODE_RUNNING = 253,
  JOINT_MODE_IDLE = 255,
}

export interface IRosTypeUrinterfacesJointTemperature {
  header: IRosTypeStdMsgsHeader;
  data: number[];
}

export interface IRosTypeUrinterfacesJointVoltage {
  header: IRosTypeStdMsgsHeader;
  data: number[];
}

export interface IRosTypeUrinterfacesRobotMode {
  header: IRosTypeStdMsgsHeader;
  data: number;
}

export enum IRosTypeUrinterfacesRobotModeConst {
  DISCONNECTED = 0,
  CONFIRM_SAFETY = 1,
  BOOTING = 2,
  POWER_OFF = 3,
  POWER_ON = 4,
  IDLE = 5,
  BACKDRIVE = 6,
  RUNNING = 7,
  UPDATING_FIRMWARE = 8,
}

export interface IRosTypeUrinterfacesRuntimeState {
  header: IRosTypeStdMsgsHeader;
  data: number;
}

export enum IRosTypeUrinterfacesRuntimeStateConst {
  PROGRAM_STATE_UNDEFINED = -1,
  PROGRAM_STATE_STOPPING = 0,
  PROGRAM_STATE_STOPPED = 1,
  PROGRAM_STATE_RUNNING = 2,
  PROGRAM_STATE_PAUSING = 3,
  PROGRAM_STATE_PAUSED = 4,
  PROGRAM_STATE_RESUMING = 5,
  PROGRAM_STATE_RETRACTING = 6,
}

export interface IRosTypeUrinterfacesSafetyStatus {
  header: IRosTypeStdMsgsHeader;
  data: number;
}

export enum IRosTypeUrinterfacesSafetyStatusConst {
  NORMAL = 1,
  REDUCED = 2,
  PROTECTIVE_STOP = 3,
  RECOVERY = 4,
  SAFEGUARD_STOP = 5,
  SYSTEM_EMERGENCY_STOP = 6,
  ROBOT_EMERGENCY_STOP = 7,
  VIOLATION = 8,
  FAULT = 9,
  VALIDATE_JOINT_ID = 10,
  UNDEFINED_SAFETY_MODE = 11,
  AUTOMATIC_MODE_SAFEGUARD_STOP = 12,
  SYSTEM_THREE_POSITION_ENABLING_STOP = 13,
}

export interface IRosTypeUrinterfacesSetAnalogOutput {
  request: IRosTypeUrinterfacesSetAnalogOutputRequest;
  response: IRosTypeUrinterfacesSetAnalogOutputResponse;
}

export interface IRosTypeUrinterfacesSetAnalogOutputRequest {
  data: IRosTypeUrinterfacesAnalog[];
}

export interface IRosTypeUrinterfacesSetAnalogOutputResponse {
  success: boolean;
}

export interface IRosTypeUrinterfacesSetDigitalOutput {
  request: IRosTypeUrinterfacesSetDigitalOutputRequest;
  response: IRosTypeUrinterfacesSetDigitalOutputResponse;
}

export interface IRosTypeUrinterfacesSetDigitalOutputRequest {
  data: IRosTypeUrinterfacesDigital[];
}

export interface IRosTypeUrinterfacesSetDigitalOutputResponse {
  success: boolean;
}

export interface IRosTypeUrinterfacesSetSpeedFraction {
  request: IRosTypeUrinterfacesSetSpeedFractionRequest;
  response: IRosTypeUrinterfacesSetSpeedFractionResponse;
}

export interface IRosTypeUrinterfacesSetSpeedFractionRequest {
  value: number;
}

export interface IRosTypeUrinterfacesSetSpeedFractionResponse {
  success: boolean;
}

export interface IRosTypeUrinterfacesSetUrProgramState {
  request: IRosTypeUrinterfacesSetUrProgramStateRequest;
  response: IRosTypeUrinterfacesSetUrProgramStateResponse;
}

export interface IRosTypeUrinterfacesSetUrProgramStateRequest {
  program_name: string;
  requested: IRosTypeUrinterfacesUrProgramStateEnum;
}

export interface IRosTypeUrinterfacesSetUrProgramStateResponse {
  success: boolean;
}

export interface IRosTypeUrinterfacesSpeedSliderFraction {
  header: IRosTypeStdMsgsHeader;
  data: number;
}

export interface IRosTypeUrinterfacesTcpState {
  header: IRosTypeStdMsgsHeader;
  pose: IRosTypeGeometryMsgsPose;
  speed: IRosTypeGeometryMsgsTwist;
}

export interface IRosTypeUrinterfacesToolOutputMode {
  header: IRosTypeStdMsgsHeader;
  output_mode: number;
  output_pin_mode: number[];
}

export enum IRosTypeUrinterfacesToolOutputModeConst {
  TOOL_OUTPUT = 0,
  TOOL_POWER = 1,
  TOOL_SAFE = 2,
  OUTPUT_MODE_DISABLED = 0,
  OUTPUT_MODE_SINKING = 1,
  OUTPUT_MODE_SOURCING = 2,
  OUTPUT_MODE_PUSH_PULL = 3,
}

export interface IRosTypeUrinterfacesUrProgramActualState {
  header: IRosTypeStdMsgsHeader;
  program_name: IRosTypeStdMsgsString;
  actual: IRosTypeUrinterfacesUrProgramStateEnum;
}

export interface IRosTypeUrinterfacesUrProgramCompiler {
  request: IRosTypeUrinterfacesUrProgramCompilerRequest;
  response: IRosTypeUrinterfacesUrProgramCompilerResponse;
}

export interface IRosTypeUrinterfacesUrProgramCompilerRequest {
  program_source_code: string;
}

export interface IRosTypeUrinterfacesUrProgramCompilerResponse {
  has_error: boolean;
  data: IRosTypeStdMsgsString;
}

export interface IRosTypeUrinterfacesUrProgramCompilerResponse {
  response: IRosTypeUrinterfacesUrProgramCompilerResponse;
}

export interface IRosTypeUrinterfacesUrProgramError {
  header: IRosTypeStdMsgsHeader;
  program_name: IRosTypeStdMsgsString;
  error_message: IRosTypeStdMsgsString;
}

export interface IRosTypeUrinterfacesUrProgramStateEnum {
  program_state: number;
}

export enum IRosTypeUrinterfacesUrProgramStateEnumConst {
  PROGRAM_STATE_UNDEFINED = -1,
  PROGRAM_STATE_STOPPING = 0,
  PROGRAM_STATE_STOPPED = 1,
  PROGRAM_STATE_RUNNING = 2,
  PROGRAM_STATE_PAUSING = 3,
  PROGRAM_STATE_PAUSED = 4,
  PROGRAM_STATE_RESUMING = 5,
}

export interface IRosTypeUrinterfacesUrThreadLabel {
  header: IRosTypeStdMsgsHeader;
  program_name: IRosTypeStdMsgsString;
  thread_id: number;
  label_string: IRosTypeStdMsgsString;
  thread_string: IRosTypeStdMsgsString;
}

export interface IRosTypeUrinterfacesVoltageCurrentStamped {
  header: IRosTypeStdMsgsHeader;
  voltage: number;
  current: number;
}

export interface IRosTypeUrinterfacesWorldModelNames {
  header: IRosTypeStdMsgsHeader;
  data: IRosTypeStdMsgsString[];
}
