// @universal-robots/urcap-utils/bin/generate-manifest-constants.js
#!/usr/bin/env node

const helper = require('./lib/manifest-helper.js');
const fs = require('fs');

const DESTINATION_FILE = './src/generated';
const EXP_CONST = "export const ";

const HEADER = "// AUTO GENERATED - DO NOT EDIT\n";
const PATH_VAR = "PATH";
const VENDOR_ID_VAR = "VENDOR_ID";
const URCAP_ID_VAR = "URCAP_ID";
const URCAP_VERSION_VAR = "URCAP_VERSION";
const URCAP_ROS_NAMESPACE_VAR = "URCAP_ROS_NAMESPACE";

function createConst(name, value, quotedValue = true) {
    if (quotedValue) {
        return EXP_CONST + name + " = '" + value + "';\n";
    } else {
        return EXP_CONST + name + " = " + value + ";\n";
    }
}

function templateExpression(varName) {
    return "${" + varName + "}";
}

function createPath(contributionId) {
    const path = "`/" +
        templateExpression(VENDOR_ID_VAR) + "/" +
        templateExpression(URCAP_ID_VAR) + "/" +
        contributionId + "`";
    return createConst(PATH_VAR, path, false);
}

function createNamespace(vendorId, urcapId) {
    const namespace = (
        "Vendor_" + vendorId + "/" +
        "URCap_" + urcapId
    ).replace(/\.|-/g, "_");
    return createConst(URCAP_ROS_NAMESPACE_VAR, namespace, true);
}

function createFileContents(urcap) {
    let fileContents = HEADER +
        createConst(VENDOR_ID_VAR, urcap.vendorId) +
        createConst(URCAP_ID_VAR, urcap.urcapId) +
        createConst(URCAP_VERSION_VAR, urcap.version) +
        createNamespace(urcap.vendorId, urcap.urcapId);
    if (urcap.webArchives.length > 0) {
        // Web-archive contribution path
        fileContents += createPath(urcap.webArchives[0].folder ? urcap.webArchives[0].folder : urcap.webArchives[0].name);
    }
    return fileContents;
}

function createFile(dstFilename1, fileContents1) {
    if (!fs.existsSync(DESTINATION_FILE)) {
        fs.mkdirSync(DESTINATION_FILE);
    }
    fs.writeFileSync(DESTINATION_FILE + "/" + dstFilename1, fileContents1, 'utf-8');

    console.log('> ' + dstFilename1 + ' created');
}

const args = process.argv.slice(2);

const manifestFile = args[0];

const urcap = helper.readManifestFile(manifestFile);
const fileContents = createFileContents(urcap);

createFile('contribution-constants.ts', fileContents);
