// @universal-robots/urcap-utils/bin/lib/manifest-helper.js
const yaml = require('js-yaml');
const fs = require('fs');
const tarHelper = require("./tar-helper");

class Urcap {
    get containers() {
        return this._containers;
    }

    set containers(value) {
        value ? this._containers = value : this._containers = []
    }

    set webArchives(archives) {
        archives ? this._webArchives = archives : this._webArchives = []
    }

    get webArchives() {
        return this._webArchives
    }

    constructor() {
        this.vendorId = ''
        this.urcapId = ''
        this.version = ''
        this._webArchives = []
        this._containers = []
    }
}

function extractURCapFromManifest(manifest) {
    const webArchives = manifest.artifacts.webArchives;
    const containers = manifest.artifacts.containers

    const result = new Urcap()
    result.vendorId = manifest.metadata.vendorID
    result.urcapId = manifest.metadata.urcapID
    result.version = manifest.metadata.version
    result.webArchives = webArchives
    result.containers = containers

    return result
}

function readManifestFile(manifestFile) {
    const manifest = yaml.load(fs.readFileSync(manifestFile, 'utf8'));
    return extractURCapFromManifest(manifest);
}

function readManifestFileFromUrcap(urcapxFile) {
    return tarHelper.readTextFileFromTar(urcapxFile, 'manifest.yaml')
        .then((text) => extractURCapFromManifest(yaml.load(text, {})))
}

module.exports = {readManifestFile, readManifestFileFromUrcap}
