// @universal-robots/ui-models/fesm2022/universal-robots-ui-models.mjs
const DialogSize = {
    TAKEOVER: 'TAKEOVER',
    XXL: 'XXL',
    XL: 'XL',
    LARGE: 'LARGE',
    DEFAULT: 'DEFAULT',
    SMALL: 'SMALL',
    XSMALL: 'XSMALL',
};
const CloseReason = {
    CANCELLED: 'CANCELLED',
    DESTROYED: 'DESTROYED',
    CONFIRMED: 'CONFIRMED',
    UNKNOWN: 'UNKNOWN',
};
const DIALOG_KEYS = {
    LOCALSTORAGE: 'ur-portal-service',
    CLOSE_EVENT: 'portalservice:close',
    UPDATE_EVENT: 'portalservice:update',
};

const SelectedInput = {
    VALUE: 'VALUE',
    VARIABLE: 'VARIABLE',
    EXPRESSION: 'EXPRESSION',
};

const getAllowedKeys = () => {
    const allowedKeys = {
        dot: '.',
        arrowDown: "ArrowDown" /* Key.ArrowDown */,
        arrowLeft: "ArrowLeft" /* Key.ArrowLeft */,
        arrowRight: "ArrowRight" /* Key.ArrowRight */,
        arrowUp: "ArrowUp" /* Key.ArrowUp */,
        dash: '-',
        plus: '+',
        tab: "Tab" /* Key.Tab */,
        backspace: "Backspace" /* Key.Backspace */,
        del: "Delete" /* Key.Delete */,
    };
    return [...Object.values(allowedKeys)];
};
var urField_model = {};

/**
 * Generated bundle index. Do not edit.
 */

export { CloseReason, DIALOG_KEYS, DialogSize, SelectedInput, getAllowedKeys };
