// @universal-robots/contribution-api/fesm2022/universal-robots-contribution-api.mjs
import { Guid } from 'guid-typescript';
import { Subject, firstValueFrom, map, merge, throwError, of } from 'rxjs';
import { finalize, tap, map as map$1, switchMap } from 'rxjs/operators';
import { Md5 } from 'ts-md5';
import { expose } from 'threads/worker';
import { registerSerializer } from 'threads';
import { SelectedInput } from '@universal-robots/ui-models';
import { g, kg, kgM, kgMPerSec, degPerSec, degPerSecSq, mPerSec, mPerSecSq, mmPerSec, mmPerSecSq, radPerSec, radPerSecSq, A, N, V, deg, m, mA, mV, mm, ms, rad, s, W, kgMSq, degreesCelsius, FixedPointNumber, converter } from '@universal-robots/utilities-units';

const UNSUBSCRIBE_MESSAGE_TYPE = 'ur.web.message.unsubscribeUpdates';
/**
 *
 */
class UREventTarget extends EventTarget {
    /**
     *
     * @param name
     */
    constructor(options) {
        super();
        this.name = options?.name || Md5.hashStr(`${options?.values?.join('')}${options?.salt}`);
    }
}
/**
 * returns a reply promise to the specified message
 *
 * @param eventTarget The worker or event target where to send the message
 * @param type The type of message being sent
 * @param data The message data
 */
function postAndAwait(eventTarget, type, data = {}) {
    const id = Guid.raw();
    return new Promise((resolve, reject) => {
        let channel;
        if (isWorker(eventTarget)) {
            channel = new MessageChannel();
            channel.port1.addEventListener('message', (event) => {
                channel?.port1.close();
                if (event.data.error) {
                    reject(event.data.error);
                }
                else {
                    resolve(event.data.data);
                }
            });
            channel.port1.start();
        }
        else {
            eventTarget.addEventListener(`reply_${type}_${id}`, (event) => {
                resolve(event.data.data);
            }, { once: true });
        }
        const message = {
            type,
            data,
            id,
        };
        const name = eventTarget?.name || self.name;
        if (name) {
            message.name = name;
        }
        dispatchWorkerOrTargetEvent(eventTarget, message, channel);
    });
}
/**
 * returns an observable that will emit any messages replied to the specified message
 *
 * @param eventTarget The worker or event target where to send the message
 * @param type The type of message being sent
 * @param data The message data
 */
function postAndObserve(eventTarget, type, data) {
    const id = Guid.raw();
    const response$ = new Subject();
    let channel;
    let callback;
    if (isWorker(eventTarget)) {
        /**
         * Handles data and sends it on to the worker through observable
         * @params the event containing data
         */
        callback = (e) => {
            const event = e;
            channel?.port1.close();
            if (event.data.error) {
                response$.error(event.data.error);
            }
            else {
                response$.next(event.data.data);
            }
        };
        channel = new MessageChannel();
        channel.port1.addEventListener('message', callback, { passive: true });
        channel.port1.start();
    }
    else {
        /**
         * Handles data and sends it on to the EventTarget through observable
         * @params the event containing data
         */
        callback = (event) => {
            if (response$.observed) {
                response$.next(event.data.data);
            }
        };
        eventTarget.addEventListener(`reply_${type}_${id}`, callback, {
            passive: true,
        });
    }
    const message = {
        type,
        data: {
            ...data,
        },
        id,
    };
    dispatchWorkerOrTargetEvent(eventTarget, message, channel);
    return response$.pipe(finalize(() => unsubscribeOnFinalized(eventTarget, type, id, callback, channel)));
}
/**
 * Sends an unsubscribe event to app for closing subscriptions and removes internal listeners
 * @param eventTarget the worker or event target where to send the message
 * @param type the type of message
 * @param id the id of the subscription
 * @param channel if sending to a worker a message channel must be provided
 * @param callback the callback to be called when unsubscribing
 */
function unsubscribeOnFinalized(eventTarget, type, id, callback, channel) {
    if (isWorker(eventTarget)) {
        channel?.port1?.removeEventListener('message', callback);
    }
    else {
        eventTarget.removeEventListener(`reply_${type}_${id}`, callback);
    }
    const unsubscribeMessage = {
        type: UNSUBSCRIBE_MESSAGE_TYPE,
        data: { id },
    };
    dispatchWorkerOrTargetEvent(eventTarget, unsubscribeMessage, channel);
}
/**
 * send a message to the specified worker or target event
 * @param eventTarget the worker or event target where to send the message
 * @param message the message to be sent
 * @param channel if sending to a worker a message channel must be provided
 */
function dispatchWorkerOrTargetEvent(eventTarget, message, channel) {
    if (isWorker(eventTarget)) {
        if (!channel?.port2) {
            throw new Error('A channel must be provided when sending to a worker');
        }
        // TODO: Don't think this is a real fix, but it is a workaround for the issue
        if (message?.type === UNSUBSCRIBE_MESSAGE_TYPE) {
            eventTarget.postMessage(message);
        }
        else {
            eventTarget.postMessage(message, [channel.port2]);
        }
    }
    else {
        eventTarget.dispatchEvent(new MessageEvent('request', { data: message }));
    }
}
/**
 * returns true if the item is a worker type
 * @param item
 */
function isWorker(item) {
    return item.postMessage && typeof item.postMessage === 'function';
}
/**
 * Abstract class for invoking messages
 */
class MessageInvoker {
    static { this.postAndAwait = postAndAwait; }
    static { this.postAndObserve = postAndObserve; }
}
/**
 * Base class for invoking messages
 */
class MessageInvokerService {
    /**
     * @constructor
     * @param messageBase
     * @param eventTarget
     */
    constructor(messageBase, eventTarget) {
        this.messageBase = messageBase;
        this.eventTarget = eventTarget;
    }
    /**
     * returns a reply promise to the specified message
     *
     * @param type The type of message being sent
     * @param data The message data
     */
    async postAndAwait(type, data = {}) {
        return MessageInvoker.postAndAwait(this.eventTarget, this.messageBase + type, data);
    }
    /**
     * returns an observable that will emit any messages replied to the specified message
     *
     * @param type The type of message being sent
     * @param data The message data
     */
    postAndObserve(type, data = {}) {
        return MessageInvoker.postAndObserve(this.eventTarget, this.messageBase + type, data);
    }
}

const APPLICATION_NODE_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.ApplicationNodeService.';
/**
 * Service that handles updating application nodes
 */
class ApplicationNodeService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        this.eventTarget = eventTarget;
    }
    /**
     * Update the program node
     * @param node New program node state
     */
    async updateNode(node) {
        return MessageInvoker.postAndAwait(this.eventTarget, APPLICATION_NODE_SERVICE_MESSAGE_TYPE_BASE + 'updateNode', node);
    }
}

/**
 * Defines the types of URSymbols
 */
var URSymbolType;
(function (URSymbolType) {
    URSymbolType["Variable"] = "$$Variable";
    URSymbolType["Function"] = "$$Function";
    URSymbolType["Module"] = "$$Module";
    URSymbolType["LogicProgram"] = "$$LogicProgram";
})(URSymbolType || (URSymbolType = {}));
/**
 * Defines abstract symbol class used by URSymbolService
 */
class URSymbol {
    /**
     *  Create new URSymbol
     * @param name
     * @param reference
     */
    constructor(name, reference = false) {
        this.name = name;
        this.reference = reference;
    }
}
/**
 * Definition of variable used for discovering variables in the tree node parameters
 */
class URVariable extends URSymbol {
    /**
     *  Create new URVariable
     * @param name
     * @param valueType
     * @param reference
     */
    constructor(name, valueType, reference = false) {
        super(name, reference);
        this.type = URSymbolType.Variable;
        this.valueType = valueType;
    }
}
/**
 * Defines the ValueTypes that can be assigned to a variable.
 */
var VariableValueType;
(function (VariableValueType) {
    VariableValueType["BOOLEAN"] = "boolean";
    VariableValueType["FLOAT"] = "float";
    VariableValueType["INTEGER"] = "integer";
    VariableValueType["POSE"] = "pose";
    VariableValueType["STRING"] = "string";
    VariableValueType["MATRIX"] = "matrix";
    VariableValueType["ARRAY"] = "array";
    VariableValueType["WAYPOINT"] = "waypoint";
    VariableValueType["GRID"] = "grid";
    VariableValueType["FRAME"] = "frame";
    VariableValueType["TIMER"] = "timer";
})(VariableValueType || (VariableValueType = {}));
/**
 * Definition of Logic Program
 */
class URLogicProgram extends URSymbol {
    constructor() {
        super(...arguments);
        this.type = URSymbolType.LogicProgram;
    }
}
/**
 * Definition of function
 */
class URFunction extends URSymbol {
    constructor() {
        super(...arguments);
        this.type = URSymbolType.Function;
    }
}
/**
 * Definition of Module
 */
class URModule extends URSymbol {
    constructor() {
        super(...arguments);
        this.type = URSymbolType.Module;
    }
}
/**
 * Check if item is a URSymbol
 * @param item
 * @returns
 */
function isURSymbol(item) {
    return (!!item &&
        Object.prototype.hasOwnProperty.call(item, 'name') &&
        Object.prototype.hasOwnProperty.call(item, 'type') &&
        Object.values(URSymbolType).includes(item.type));
}
/**
 * TypeGuard to verify whether a given object is a variable
 * @param item The item to be interrogated
 */
function isVariable(item) {
    return isURSymbol(item) && item.type === URSymbolType.Variable;
}
/**
 * TypeGuard to verify whether a given object is a function
 * @param item The item to be interrogated
 */
function isFunction(item) {
    return isURSymbol(item) && item.type === URSymbolType.Function;
}
/**
 * TypeGuard to verify whether a given object is a module
 * @param item The item to be interrogated
 */
function isModule(item) {
    return isURSymbol(item) && item.type === URSymbolType.Module;
}
/**
 * TypeGuard to verify whether a given object is a logic program
 * @param item The item to be interrogated
 */
function isLogicProgram(item) {
    return isURSymbol(item) && item.type === URSymbolType.LogicProgram;
}

/** Describes the various axes used to define poses in the application. */
/** Position */
/**  Constant array of the position axes. */
const PositionAxes = ['x', 'y', 'z'];
/** Orientation */
/** Constant array of the orientation axes. */
const OrientationAxes = ['rx', 'ry', 'rz'];
/** Center of Gravity */
/** Constant array of the center of gravity axes. */
const CenterOfGravityAxes = ['cx', 'cy', 'cz'];
/** Pose */
/** Constant array of the combination of a position and an orientation into a pose. */
const PoseAxes = [...PositionAxes, ...OrientationAxes];

/** Position */
/**  Constant array of the InertiaMatrixKeys. */
const InertiaMatrixKeys = ['Ixx', 'Iyy', 'Izz', 'Ixy', 'Ixz', 'Iyz'];

/**
 * Checks if the value implements ReferenceItem
 */
const isReferenceItem = (value) => {
    return value.type === 'Reference';
};
/**
 * Checks if the value implements ValueItem
 */
const isValueItem = (value) => {
    return value.type === 'Value';
};

/**
 * A type guard to check if an object is a JointPositions
 */
function isJointPositions(jp) {
    return (jp.base !== undefined &&
        jp.shoulder !== undefined &&
        jp.elbow !== undefined &&
        jp.wrist1 !== undefined &&
        jp.wrist2 !== undefined &&
        jp.wrist3 !== undefined);
}

/**
 * A type guard to check if an object is a KinematicPosition
 */
function iskinematicPosition(kinematicPosition) {
    return (kinematicPosition?.jointVector !== undefined &&
        kinematicPosition?.links !== undefined &&
        kinematicPosition?.tcp !== undefined);
}

/**
 * OperatorInputType enum
 */
var OperatorInputType;
(function (OperatorInputType) {
    OperatorInputType["INPUT"] = "input";
    OperatorInputType["MESSAGE"] = "message";
})(OperatorInputType || (OperatorInputType = {}));

/**
 * PopupLevel enum
 */
var PopupLevel;
(function (PopupLevel) {
    PopupLevel["ERROR"] = "error";
    PopupLevel["WARNING"] = "warning";
    PopupLevel["INFO"] = "info";
})(PopupLevel || (PopupLevel = {}));

/**
 * A type guard to check if an object is a Waypoint
 */
function isWaypoint(waypoint) {
    return (waypoint?.frame !== undefined &&
        waypoint?.pose !== undefined &&
        waypoint?.qNear !== undefined);
}

/**
 * An enum of all the built in application node types
 */
var ApplicationNodeType;
(function (ApplicationNodeType) {
    ApplicationNodeType["MOUNTING"] = "ur-mounting";
    ApplicationNodeType["FRAMES"] = "ur-frames";
    ApplicationNodeType["GRID_PATTERN"] = "ur-grid-pattern";
    ApplicationNodeType["END_EFFECTOR"] = "ur-end-effector";
    ApplicationNodeType["SMART_SKILLS"] = "ur-smart-skills";
    ApplicationNodeType["MOTION_PROFILES"] = "ur-motion-profiles";
})(ApplicationNodeType || (ApplicationNodeType = {}));

var SmartSkillType;
(function (SmartSkillType) {
    SmartSkillType["ALIGN_TO_Z"] = "ur-align-to-z";
    SmartSkillType["ALIGN_TO_PLANE"] = "ur-align-to-plane";
    SmartSkillType["ALIGN_Z_TO_NEAREST_AXIS"] = "ur-align-z-to-nearest-axis";
    SmartSkillType["CENTER"] = "ur-center";
    SmartSkillType["FREEDRIVE"] = "ur-freedrive";
    SmartSkillType["MOVE_INTO_CONTACT"] = "ur-move-into-contact";
    SmartSkillType["RETRACT"] = "ur-retract";
    SmartSkillType["PUT_IN_BOX"] = "ur-put-in-box";
    SmartSkillType["CUSTOM_SMART_SKILL"] = "ur-custom-smart-skill";
    SmartSkillType["ADMITTANCE_CONTROL"] = "ur-admittance-control";
})(SmartSkillType || (SmartSkillType = {}));

var MotionProfileMoveType;
(function (MotionProfileMoveType) {
    MotionProfileMoveType["Joint"] = "joint";
    MotionProfileMoveType["Linear"] = "linear";
    MotionProfileMoveType["Process"] = "process";
})(MotionProfileMoveType || (MotionProfileMoveType = {}));
var MovementType;
(function (MovementType) {
    MovementType["Optimove"] = "OptiMove";
    MovementType["Classic"] = "Classic";
})(MovementType || (MovementType = {}));

const floatOperators = ['<', '>'];
const registerOperators = [...floatOperators, '==', '!=', '>=', '<='];
const operatorInverseMap = {
    '<': '>=',
    '>': '<=',
    '==': '!=',
    '!=': '==',
    '>=': '<',
    '<=': '>',
};

var WaypointType;
(function (WaypointType) {
    WaypointType["Teach"] = "teach";
    WaypointType["Expression"] = "expression";
})(WaypointType || (WaypointType = {}));

var NodeType;
(function (NodeType) {
    NodeType["PROGRAM"] = "ur-program";
    NodeType["FUNCTIONS"] = "ur-functions";
    NodeType["CODE"] = "ur-code";
    NodeType["BEFORE_START"] = "ur-before-start";
    NodeType["CONFIGURATION"] = "ur-configuration";
    NodeType["MODULES"] = "ur-modules";
    NodeType["STATUS"] = "ur-status";
    NodeType["FUNCTION"] = "ur-function";
    NodeType["MODULE"] = "ur-module";
    NodeType["WAIT"] = "ur-wait";
    NodeType["LOOP"] = "ur-loop";
    NodeType["ASSIGNMENT"] = "ur-assignment";
    NodeType["FOLDER"] = "ur-folder";
    NodeType["SCRIPT"] = "ur-script";
    NodeType["MOVE"] = "ur-move";
    NodeType["IF"] = "ur-if";
    NodeType["ELSE"] = "ur-else";
    NodeType["ELSEIF"] = "ur-elseif";
    NodeType["WAYPOINT"] = "ur-waypoint";
    NodeType["SET"] = "ur-set";
    NodeType["COMMENT"] = "ur-comment";
    NodeType["TOOL_FORCE"] = "ur-tool-force";
    NodeType["OPERATOR_INPUT"] = "ur-operator-input";
    NodeType["SET_TCP"] = "ur-set-tcp";
    NodeType["CALL_FUNCTION"] = "ur-call-function";
    NodeType["STATUS_ITEM"] = "ur-status-item";
    NodeType["GRID_ITERATOR"] = "ur-grid-iterator";
    NodeType["CONFIGURATION_VALUE_ITEM"] = "ur-configuration-value-item";
    NodeType["CONFIGURATION_REFERENCE_ITEM"] = "ur-configuration-reference-item";
    NodeType["CONFIGURATION_REFERENCE_ITEM_OPTION"] = "ur-configuration-reference-item-option";
    NodeType["MOVE_TO"] = "ur-move-to";
    NodeType["JOINT_MOVE"] = "ur-joint-move";
    NodeType["LINEAR_MOVE"] = "ur-linear-move";
    NodeType["PATH"] = "ur-path";
    NodeType["PAYLOAD"] = "ur-payload";
    NodeType["TCP"] = "ur-tcp";
    NodeType["ALIGN_TO"] = "ur-align-to";
    NodeType["THREAD"] = "ur-thread";
    NodeType["LOGIC_PROGRAMS"] = "ur-logic-programs";
    NodeType["LOGIC_PROGRAM"] = "ur-logic-program";
})(NodeType || (NodeType = {}));

const ThreadAction = {
    KILL: 'Kill',
    RUN: 'Run',
};

const APPLICATION_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.ApplicationService.';
/**
 * Provides access to data stored in the application
 */
class ApplicationService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(APPLICATION_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Returns a list of the Application Nodes that contribute data to the currently loaded Application
     */
    getAvailableApplicationNodes() {
        return this.postAndAwait('getAvailableApplicationNodes');
    }
    /**
     * Returns the data stored on the specified Application Node
     * @param name the name of the Application Node to fetch
     */
    getApplicationNode(name) {
        return this.postAndAwait('getApplicationNode', { name });
    }
    /**
     * Returns true if a grid from application is valid. If the variable name is not a grid
     * or a grid defined somewhere else then it always returns true.
     * @param name The name of the variable being validated
     * @returns true if the grid is valid or is not a grid but a different variable type
     */
    async isValidGridPattern(name) {
        const gridPatternNode = (await this.getApplicationNode(ApplicationNodeType.GRID_PATTERN));
        const grid = gridPatternNode.grids.find((grid) => grid.grid.name === name);
        if (!grid) {
            // Grid is defined somewhere else, so we cannot validate it and assume it is valid
            return true;
        }
        // A grid pattern needs to have at least some corner positions defined.
        if (!grid.corners.length) {
            return false;
        }
        // Determine whether the grid pattern is 1D (single row or column) or 2D (matrix).
        // 1D grids need at least 2 corners.
        // 2D grids need at least 4 corners.
        const columns = grid.numColumns;
        const rows = grid.numRows;
        const requiredCorners = columns === 1 || rows === 1 ? 2 : 4;
        const hasRequiredCorners = grid.corners.length >= requiredCorners;
        // Every corner must be defined to be a valid grid.
        const hasValidCorners = grid.corners.every((corner) => !!corner);
        return hasRequiredCorners && hasValidCorners;
    }
}

const DIALOG_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.DialogService.';
/**
 * Service for open a dialog
 */
class DialogService {
    /**
     * Create a service for the specific worker
     * @param worker
     */
    constructor(worker) {
        this.worker = worker;
    }
    /**
     * Open a Confirmation Dialog with a single dismiss button.
     * @param title Dialog title
     * @param text Dialog message
     * @param icon Dialog icon. It can be 'info', 'warning', 'error', or other icon strings.
     * @param confirmText Optional. Text for confirm button. Default to 'OK'.
     * @param  confirmIcon Optional. icon for confirm button. Default to no icon.
     */
    openConfirmDialog(title, text, icon, confirmText, confirmIcon) {
        return MessageInvoker.postAndAwait(this.worker, DIALOG_SERVICE_MESSAGE_TYPE_BASE + 'openConfirmDialog', {
            title,
            text,
            icon,
            textIsTranslationKey: false,
            confirmText,
            confirmIcon,
            replacementPatterns: undefined,
            replacements: undefined,
            showCancelButton: false,
            cancelText: undefined,
        });
    }
    /**
     * Open a Confirmation Dialog with a single dismiss button.
     * @param title Dialog title
     * @param text Dialog message
     * @param icon Dialog icon. It can be 'info', 'warning', 'error', or other icon strings.
     * @param confirmText Optional. Text for confirm button. Default to 'Confirm'.
     * @param confirmIcon Optional. Icon for confirm button. Default to no icon.
     * @param cancelText Optional. Text for cancel button. Default to 'Go Back'.
     */
    openConfirmCancelDialog(title, text, icon, confirmText, confirmIcon, cancelText) {
        return MessageInvoker.postAndAwait(this.worker, DIALOG_SERVICE_MESSAGE_TYPE_BASE + 'openConfirmDialog', {
            title,
            text,
            icon,
            textIsTranslationKey: false,
            confirmText,
            confirmIcon,
            replacementPatterns: undefined,
            replacements: undefined,
            showCancelButton: true,
            cancelText,
        });
    }
    /**
     * Open a Custom Webcomponent Dialog
     * @param componentTag The tagname of the webcomponent to render in the dialog
     * @param initialData The initial data to be passed to the dialog component upon opening
     * @param options
     */
    openCustomDialog(componentTag, initialData, options) {
        return MessageInvoker.postAndAwait(this.worker, DIALOG_SERVICE_MESSAGE_TYPE_BASE + 'openCustomDialog', {
            componentTag,
            inputData: initialData,
            options,
        });
    }
}

const LOG_MESSAGE_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.LogMessageService.';
/**
 * Service for sending log messages
 */
class LogMessageService {
    /**
     * Create a service for the specific worker
     * @param worker
     */
    constructor(worker) {
        this.worker = worker;
    }
    /**
     * Sends a log message.
     *
     * @param message The message to be logged
     */
    sendLogMessage(message) {
        return MessageInvoker.postAndAwait(this.worker, LOG_MESSAGE_SERVICE_MESSAGE_TYPE_BASE + 'sendLogMessage', message);
    }
    /**
     * Sends a warning log message.
     *
     * @param message The message to be logged
     * @param type The type of message - if not specified 'generic' will be used
     */
    warn(message, type = 'generic') {
        return MessageInvoker.postAndAwait(this.worker, LOG_MESSAGE_SERVICE_MESSAGE_TYPE_BASE + 'warn', { message, type });
    }
    /**
     * Sends a info log message.
     *
     * @param message The message to be logged
     * @param type The type of message - if not specified 'generic' will be used
     */
    info(message, type = 'generic') {
        return MessageInvoker.postAndAwait(this.worker, LOG_MESSAGE_SERVICE_MESSAGE_TYPE_BASE + 'info', { message, type });
    }
    /**
     * Sends a debug log message.
     *
     * @param message The message to be logged
     * @param type The type of message - if not specified 'generic' will be used
     */
    debug(message, type = 'generic') {
        return MessageInvoker.postAndAwait(this.worker, LOG_MESSAGE_SERVICE_MESSAGE_TYPE_BASE + 'debug', { message, type });
    }
    /**
     * Sends a error log message.
     *
     * @param message The message to be logged
     * @param type The type of message - if not specified 'generic' will be used
     */
    error(message, type = 'generic') {
        return MessageInvoker.postAndAwait(this.worker, LOG_MESSAGE_SERVICE_MESSAGE_TYPE_BASE + 'error', { message, type });
    }
}

const LOG_PERFORMANCE_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.LogPerformanceService.';
/**
 * Service for sending log performance metrics
 */
class LogPerformanceService extends MessageInvokerService {
    /**
     * Create a service for the specific worker
     * @param worker
     */
    constructor(worker) {
        super(LOG_PERFORMANCE_SERVICE_MESSAGE_TYPE_BASE, worker);
    }
    /**
     * Sends a basic log performance metric where duration, start and end names are the same.
     *
     * @param markName
     */
    logPerformanceMetrics(markName) {
        return this.postAndAwait('logPerformanceMetrics', markName);
    }
    /**
     * Sends a extended log performance metric where duration, start and end names can be specified individually.
     *
     * @param durationName The duration name
     * @param startName The start name
     * @param endName The end name
     */
    logPerformanceMetricsExtended(durationName, startName, endName) {
        return this.postAndAwait('logPerformanceMetricsExtended', { durationName, startName, endName });
    }
}

const PROGRAM_NODE_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.ProgramNodeService.';
/**
 * Service that handles node insertion and removal og nodes
 */
class ProgramNodeService {
    /**
     * @constructor
     * @param eventTarget
     * @param selectedNodeId
     */
    constructor(eventTarget, selectedNodeId) {
        this.eventTarget = eventTarget;
        this.selectedNodeId = selectedNodeId;
    }
    /**
     * Update the program node
     * @param node New program node state
     */
    async updateNode(node) {
        return MessageInvoker.postAndAwait(this.eventTarget, PROGRAM_NODE_SERVICE_MESSAGE_TYPE_BASE + 'updateParentNode', {
            node,
            id: this.selectedNodeId,
        });
    }
}

const ROBOT_MOVE_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.RobotMoveService.';
/**
 * RobotMoveService handles movement screen navigation
 */
class RobotMoveService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(ROBOT_MOVE_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Opens the move screen and returns the selected position of the robot. Returns undefined if cancel is selected
     * @param options Move screen options
     */
    async openMoveScreen(options) {
        return this.postAndAwait('openMoveScreen', options);
    }
    /**
     * Opens the autoMove screen
     * @param targetPosition Joint vector to move to
     */
    async autoMove(targetPosition) {
        return this.postAndAwait('autoMove', targetPosition);
    }
}

const ROBOT_POSITION_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.RobotPositionService.';
/**
 * RobotPositionService is used to obtain information on the current robot position
 */
class RobotPositionService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(ROBOT_POSITION_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Gets the id of the currently selected frame id to perform moves relative to
     */
    async getRelativeToFrameId() {
        return this.postAndAwait('getRelativeToFrameId');
    }
    /**
     * Gets the kinematic info
     */
    async getKinematicInfo() {
        return this.postAndAwait('getKinematicInfo');
    }
    /**
     * Get the inverse kinematics for the given pose and qNear
     * @param pose The pose to calculate the inverse kinematics for
     * @param qNear The joint positions to start the search from
     */
    async getInverseKinematics(pose, qNear) {
        return this.postAndAwait('getInverseKinematics', {
            pose,
            qNear,
        });
    }
    /**
     * Get the current pose of the TCP relative to the given frame
     */
    getPoseRelativeToFrame(frame) {
        return this.postAndObserve('getPoseRelativeToFrame', { frame });
    }
    /**
     * Get the current joint positions of the robot
     */
    getJointPositions() {
        return this.postAndObserve('getJointPositions');
    }
    /**
     * Get current position of the robot
     */
    async getCurrentPosition() {
        return this.postAndAwait('getCurrentPosition');
    }
}

const SAFETY_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.SafetyService.';
/**
 * Service for getting information about Safety
 */
class SafetyService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(SAFETY_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Returns a list of safety IO's
     */
    getIOs() {
        return this.postAndAwait('getIOs');
    }
    /**
     * Get the maximum allowed tool force for the robot
     */
    getMaxToolForce() {
        return this.postAndAwait('getMaxToolForce');
    }
    /**
     * Get the payload limits for the robot
     */
    getPayloadLimits() {
        return this.postAndAwait('getPayloadLimits');
    }
    /**
     * Get the operational mode of the robot
     */
    getOperationalMode() {
        return this.postAndObserve('getOperationalMode');
    }
}

const SOURCE_MESSAGE_TYPE_BASE = 'ur.web.message.SourceService.';
/**
 * Service for getting information about Sources and Signals
 */
class SourceService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        this.eventTarget = eventTarget;
    }
    /**
     * Returns the ID's of Sources that are available to the application
     * @param filter SignalFilter used to filter only Sources which contain Signals matching the filter
     */
    getSources(filter) {
        return firstValueFrom(MessageInvoker.postAndObserve(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'sourcesList', { filter }).pipe(map((sources) => {
            return sources.map((source) => source.sourceID);
        })));
    }
    /**
     * Returns the map of Sources
     */
    sources(filter) {
        return MessageInvoker.postAndObserve(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'sources', { filter });
    }
    /**
     * Returns the source node label for the griven groupId
     */
    sourceNodeLabel(groupId) {
        return this.sourceNodeLabels().pipe(map((labels) => labels[groupId]));
    }
    /**
     * Returns the map of source node labels
     */
    sourceNodeLabels() {
        return MessageInvoker.postAndObserve(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getSourceNodeLabels', {});
    }
    /**
     * Returns the list of Signals in a Source
     * @param sourceID The Source to list Signals for
     * @param filter SignalFilter used to filter only Signals which match the filter
     * @deprecated use getSourceSignals instead
     */
    getSignals(sourceID, filter) {
        return firstValueFrom(MessageInvoker.postAndObserve(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getSourceSignals', { sourceID, filter }));
    }
    /**
     * Returns the list of Signals in a Source
     * @param sourceGroupID The Source group, containing the source to provide signals for
     * @param sourceID The Source to list Signals for
     * @param filter SignalFilter used to filter only Signals which match the filter
     */
    sourceSignals(sourceGroupID, sourceID, filter) {
        return MessageInvoker.postAndObserve(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getSourceSignals', {
            sourceGroupID,
            sourceID,
            filter,
        });
    }
    /**
     * Returns an Observable providing Signal events
     * @param sourceID The Source containing the Signals to provide events for
     * @param filter SignalFilter used to filter only events for Signals which match the filter
     * @deprecated use getSourceUpdates instead
     */
    getSignalUpdates(sourceID, filter) {
        return MessageInvoker.postAndObserve(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getSignalUpdates', { sourceID, filter });
    }
    /**
     * Returns an Observable providing Signal events
     * @param sourceGroupID The Source group, containing the source to provide events for
     * @param sourceID The Source containing the Signals to provide events for
     * @param filter SignalFilter used to filter only events for Signals which match the filter
     */
    getSourceUpdates(sourceGroupID, sourceID, filter) {
        // Combine results of "old" and "new" APIs, and return whichever produces a value
        // The old message
        const o1 = MessageInvoker.postAndObserve(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getSignalUpdates', {
            sourceID,
            filter,
        });
        // The new message
        const o2 = MessageInvoker.postAndObserve(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getSourceUpdates', {
            sourceGroupID,
            sourceID,
            filter,
        });
        return merge(o1, o2);
    }
    /**
     * Returns the generated script code for setting the signal to the given value
     * @param sourceID The ID of the source containing the signal to set
     * @param signalID The ID of the signal to set
     * @param value The value to set the signal to - for digital values use 0 or 1, for registers use only integers, for analog values use any float
     * @deprecated use getSetSignalScript instead
     */
    generateSetSignalScript(sourceID, signalID, value) {
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'generateSetSignalScript', {
            sourceID,
            signalID,
            value,
        });
    }
    /**
     * Returns the generated script code for setting the signal to the given value
     * @param sourceGroupID The Source group, containing the source and signal to provide script for
     * @param sourceID The ID of the source containing the signal to set
     * @param signalID The ID of the signal to set
     * @param value The value to set the signal to - for digital values use 0 or 1, for registers use only integers, for analog values use any float
     */
    getSetSignalScript(sourceGroupID, sourceID, signalID, value) {
        // Combine results of "old" and "new" APIs, and return whichever produces a value
        return new Promise((resolve, reject) => {
            Promise.all([
                // The old message
                MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'generateSetSignalScript', {
                    sourceID,
                    signalID,
                    value,
                }),
                // The new message
                MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getSetSignalScript', {
                    sourceGroupID,
                    sourceID,
                    signalID,
                    value,
                }),
            ])
                .then((value) => {
                // Return whichever value (if any) that comes back (priority on the "new" value)
                resolve(value[1] ? value[1] : value[0] ? value[0] : '');
            })
                .catch((reason) => reject(reason));
        });
    }
    /**
     * Returns the generated script code for getting the value of a given signal
     * @param sourceID The ID of the source containing the signal to get
     * @param signalID The ID of the signal to get
     * @deprecated use getGetSignalScript instead
     */
    generateGetSignalScript(sourceID, signalID) {
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'generateGetSignalScript', {
            sourceID,
            signalID,
        });
    }
    /**
     * Returns the generated script code for getting the value of a given signal
     * @param groupId The Source group, containing the source to provide events for
     * @param sourceId The ID of the source containing the signal to get
     * @param signalId The ID of the signal to get
     */
    getGetSignalScript(groupId, sourceId, signalId) {
        // Combine results of "old" and "new" APIs, and return whichever produces a value
        return new Promise((resolve, reject) => {
            Promise.all([
                // The old message
                MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'generateGetSignalScript', {
                    sourceID: sourceId,
                    signalID: signalId,
                }),
                // The new message
                MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getGetSignalScript', {
                    groupId,
                    sourceId,
                    signalId,
                }),
            ])
                .then((value) => {
                // Return whichever value (if any) that comes back (priority on the "new" value)
                resolve(value[1] ? value[1] : value[0] ? value[0] : '');
            })
                .catch((reason) => reject(reason));
        });
    }
    /**
     * Returns an object literal containing a mapping of SignalIDs -> AnalogDomains
     * @param sourceID The ID of the Source for which to get Domain data
     * @deprecated should be handled by the individual contribution
     */
    getAnalogSignalDomains(sourceID) {
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getAnalogSignalDomains', { sourceID });
    }
    /**
     * Performs validation on a new value that will be used in a set script
     * @param groupId The contribution groupId
     * @param sourceId The contribution sourceId
     * @param signalId The contribution signalId
     * @param value The value that will be set
     */
    validateSetSignal(groupId, sourceId, signalId, value) {
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'validateSetSignal', {
            groupId,
            sourceId,
            signalId,
            value,
        });
    }
    /**
     * Performs validation on a new value that will be used in a get script
     * @param groupId The contribution groupId
     * @param sourceId The contribution sourceId
     * @param signalId The contribution signalId
     * @param value The value that will be got
     * @param operator The operator on the value that will be got
     */
    validateGetSignal(groupId, sourceId, signalId, value, operator) {
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'validateGetSignal', {
            groupId,
            sourceId,
            signalId,
            value,
            operator,
        });
    }
    /**
     * Returns info about the signals for a given source
     * @param groupId
     * @param sourceId
     */
    getSignalsDomainData(groupId, sourceId) {
        // What this needs to do is:
        // If it's the old tool/wired IO (if the groupId is 'ur-robot') then call through to the getSignalsDomainData method
        // on the data-store SourceService
        // Else call the getSignalsDomainData method from the behavior of the groupId SourcesNode
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getSignalsDomainData', { groupId, sourceId });
    }
    /**
     * Sets a value on a signal
     * @param sourceID The ID of the Source, containing the Signal to set
     * @param signalID The ID of the Signal to set
     * @param value The value to set
     * @deprecated use setSourceSignalValue instead
     */
    setSignalValue(sourceID, signalID, value) {
        const updateSignalValue = {
            sourceID,
            signalID,
            value,
        };
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'setSignalValue', updateSignalValue);
    }
    /**
     * Sets a value on a signal
     * @param sourceGroupID The Source group, containing the source to provide events for
     * @param sourceID The ID of the Source, containing the Signal to set
     * @param signalID The ID of the Signal to set
     * @param value The value to set
     */
    setSourceSignalValue(sourceGroupID, sourceID, signalID, value) {
        const updateSignalValue = {
            sourceID,
            signalID,
            value,
        };
        const setSignalValue = {
            sourceGroupID,
            sourceID,
            signalID,
            value,
        };
        // Send the update to both "old" and "new" APIs
        return new Promise((resolve, reject) => {
            Promise.all([
                // The old message
                MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'setSignalValue', updateSignalValue),
                // The new message
                MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'setSourceSignalValue', setSignalValue),
            ])
                .then(() => resolve())
                .catch((reason) => {
                reject(reason);
            });
        });
    }
    /**
     * Returns an object literal containing a mapping of SignalIDs -> Labels
     * @param sourceID The ID of the Source for which to get the Label mappings
     * @deprecated Source contributions will be responsible for providing labels for signals directly
     */
    getSignalLabels(sourceID) {
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getSignalLabels', sourceID);
    }
    /**
     * Returns an object literal containing a mapping of SignalIDs -> Presets
     * @param sourceID The ID of the Source for which to get the Preset mappings
     * @deprecated Source contributions will be responsible for handling presets themselves
     */
    getSignalPresets(sourceID) {
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'getSignalPresets', sourceID);
    }
}

/**
 *
 */
class SignalService {
    /**
     *
     */
    constructor(eventTarget, sourceID) {
        this.eventTarget = eventTarget;
        this.sourceID = sourceID;
    }
    /** Update the label for a specific signal id
     * @param signalID id for the signal being labelled
     * @param label the new label for the signal
     * @deprecated Source contributions will be responsible for labels for signals directly
     */
    async updateSignalLabel(signalID, label) {
        const setLabel = {
            sourceID: this.sourceID,
            signalID,
            label,
        };
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'updateSignalLabel', setLabel);
    }
    /** Update the Analog Domain for a specific signal id
     * @param signalID id for the signal being changed
     * @param analogDomainValueEnum the domain to set
     * @deprecated should be handled by the individual contribution
     */
    async updateSignalAnalogDomain(signalID, analogDomainValueEnum) {
        const updateAnalogDomainValue = {
            sourceID: this.sourceID,
            signalID,
            analogDomain: analogDomainValueEnum,
        };
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'updateSourceSignalAnalogDomain', updateAnalogDomainValue);
    }
    /**
     * Set the Preset Action for a given Digital-In signal
     * @param signalID ID of the signal to set the preset for
     * @param preset the preset to set
     * @deprecated Source contributions will be responsible for handling presets themselves
     */
    setDigitalInPreset(signalID, preset) {
        const updateAnalogInPreset = {
            sourceID: this.sourceID,
            signalID,
            preset,
        };
        return MessageInvoker.postAndAwait(this.eventTarget, SOURCE_MESSAGE_TYPE_BASE + 'setDigitalInPreset', updateAnalogInPreset);
    }
}

const SOURCE_NODE_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.SourceNodeService.';
/**
 * Service that handles updating application nodes
 */
class SourceNodeService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(SOURCE_NODE_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Update the source node
     * @param node New source node state
     */
    updateNode(node) {
        return this.postAndAwait('updateNode', node);
    }
}

const TREE_BUILDER_MESSAGE_TYPE_BASE = 'ur.web.message.TreeBuilder.';
/**
 * TreeBuilder
 *  Provides methods to build program node trees that can then be passed to
 *  the application for insertion to the main program.
 */
class TreeBuilder {
    /**
     * @constructor
     */
    constructor(worker) {
        this.worker = worker;
    }
    /**
     * Turns the passed in ProgramNode into a BranchNode, allowing children to
     * be added to the subtree of the given node.
     * @param node The ProgramNode to convert to a BranchNode
     */
    createBranchNode(node) {
        return {
            children: [],
            type: node.type,
            version: node.version,
            allowsChildren: node.allowsChildren,
            lockChildren: node.lockChildren,
            parameters: node.parameters,
        };
    }
    /**
     * Appends a child node to the subtree of the given parent node
     * @param parentNode The BranchNode to add the child to
     * @param node The child ProgramNode to be added to the subtree
     */
    appendChild(parentNode, node) {
        parentNode.children.push(node);
    }
    /**
     * Requests a new instance of a ProgramNode from the application
     * @param ofType The type of the ProgramNode to be instantiated - see the
     * nodeType enum for the builtin types.
     */
    async createNode(ofType) {
        return MessageInvoker.postAndAwait(this.worker, TREE_BUILDER_MESSAGE_TYPE_BASE + 'createNode', ofType);
    }
}

const SYMBOL_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.SymbolService.';
/**
 * Service for generating and getting reserved variable names
 */
class SymbolService {
    /**
     * Create a service for the specific worker
     * @param worker
     */
    constructor(worker) {
        this.worker = worker;
    }
    /**
     * Get a list of available variables
     * @returns
     */
    getVariables() {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'getVariables', {});
    }
    /**
     * Get an observable of available variables
     */
    variables() {
        return MessageInvoker.postAndObserve(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'variables', {});
    }
    /**
     * Returns a list of the global variables as a stream.
     * @returns {Observable<Array<GlobalVariable>>} Observable of global variables object
     */
    globalVariables() {
        return MessageInvoker.postAndObserve(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'globalVariables', {});
    }
    /**
     * Generates a new variable. The variable name is based on the suggested name.
     * If the suggested name is not usable a new, based on the suggestion, will be created
     *
     * @param suggestedName The intended name for the variable, will be suffixed if the name is already taken
     * @param valueType The ValueType of the variable
     * @param blacklist Optional parameter to provide additional names which shouldn't be used.
     */
    generateVariable(suggestedName, valueType, blacklist) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'generateVariable', {
            suggestedName,
            valueType,
            blacklist,
        });
    }
    /**
     * Return true if the provided name is the name of a registered variable
     *
     * @param variableName
     */
    isRegisteredVariableName(variableName) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'isRegisteredVariableName', variableName);
    }
    /**
     * Get a list of available functions for the given module name. If no name provided, functions for the global scope
     * @param moduleName
     * @returns
     */
    getFunctions(moduleName = '') {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'getFunctions', { moduleName });
    }
    /**
     * Get functions map. Contains a list of functions for each module name. Note: Empty string is the global scope.
     * @returns
     */
    getFunctionsMap() {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'getFunctionsMap', {});
    }
    /**
     * Generates a new function. The function name is based on the suggested name.
     * If the suggested name is not usable a new, based on the suggestion, will be created
     *
     * @param suggestedName
     * @param moduleName
     */
    generateFunction(suggestedName, moduleName) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'generateFunction', {
            functionName: suggestedName,
            moduleName,
        });
    }
    /**
     * Return true if the provided name is the name of a registered function
     *
     * @param functionName
     * @param moduleName
     */
    isRegisteredFunctionName(functionName, moduleName) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'isRegisteredFunctionName', {
            functionName,
            moduleName,
        });
    }
    /**
     * Get a list of available modules
     * @returns
     */
    getModuleNames() {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'getModuleNames', {});
    }
    /**
     * Generates a new module. The module name is based on the suggested name.
     * If the suggested name is not usable a new, based on the suggestion, will be created
     *
     * @param suggestedName
     */
    generateModule(suggestedName) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'generateModule', suggestedName);
    }
    /**
     * Return true if the provided name is the name of a registered module
     *
     * @param moduleName
     */
    isRegisteredModuleName(moduleName) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'isRegisteredModuleName', { moduleName });
    }
    /**
     * Get a list of available Logic Programs
     * @returns
     */
    getLogicProgramNames() {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'getLogicProgramNames', {});
    }
    /**
     * Generates a new Logic Program. The program name is based on the suggested name.
     * If the suggested name is not usable a new, based on the suggestion, will be created
     *
     * @param suggestedName
     */
    generateLogicProgram(suggestedName) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'generateLogicProgram', suggestedName);
    }
    /**
     * Return true if the provided name is the name of a registered Logic Program
     *
     * @param logicProgramName
     */
    isRegisteredLogicProgramName(logicProgramName) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'isRegisteredLogicProgramName', {
            logicProgramName,
        });
    }
    /**
     *  Return true if the provided name is a registered name of the given type.
     * @param name
     * @param type
     * @param moduleName
     * @param unique
     * @returns
     */
    isRegisteredName(name, type, moduleName, unique = false) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'isRegisteredName', {
            name,
            type,
            moduleName,
            unique,
        });
    }
    /**
     * Return true if the declaring program node is suppressed
     * @param name The name of the symbol
     * @param moduleName The module of the symbol if defined
     * @return Boolean describing if the declaring node is suppressed
     */
    isSuppressed(name, moduleName) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'isSuppressed', { name, moduleName });
    }
    /**
     * Get the URVariable with the given name
     * @return URVariable with the given name
     * @param name Name of the variable to get
     */
    getVariable(name) {
        return MessageInvoker.postAndAwait(this.worker, SYMBOL_SERVICE_MESSAGE_TYPE_BASE + 'getVariable', { name });
    }
    /**
     *
     * @param variableName name of the variable being validated
     * @param acceptedTypes an optional array of valid type strings, e.g., float.
     * @returns true if the variable is registered not suppressed, and is of a valid value type
     */
    async isValidVariable(variableName, acceptedTypes) {
        const isRegistered = await this.isRegisteredVariableName(variableName);
        const isSuppressed = await this.isSuppressed(variableName);
        const variable = await this.getVariable(variableName);
        let isOfValidValueType = true;
        if (acceptedTypes) {
            isOfValidValueType = acceptedTypes.some((acceptedType) => acceptedType === variable?.valueType);
        }
        return isRegistered && !isSuppressed && isOfValidValueType;
    }
}

const VALIDATION_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.ValidationService.';
/**
 * Allows program nodes to obtain the validation response generated from their respective behavior worker.
 */
class ValidationService extends MessageInvokerService {
    /**
     * Creates a new instance of the ValidationService
     * @param eventTarget the target to send message events on
     * @param selectedNodeId the id of the currently selected node
     */
    constructor(eventTarget, selectedNodeId) {
        super(VALIDATION_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
        this.selectedNodeId = selectedNodeId;
    }
    /**
     * Returns the last validation response generated from the behavior worker.
     * If the node is valid the returned response will be exactly { isValid: true } without any extra data.
     */
    async getValidationResponse(nodeId = this.selectedNodeId) {
        return this.postAndAwait('getValidationResponse', {
            id: nodeId,
        });
    }
}

const SNACKBAR_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.SnackbarService.';
/**
 * Service that handles showing snack notifications.
 */
class SnackbarService {
    /**
     * Create a service for the specific worker
     * @param worker
     */
    constructor(worker) {
        this.worker = worker;
    }
    /**
     * Open a snack notification.
     * @param message Snackbar message
     * @param type Optional. Type of notification (success, error, warning, info(default)).
     * @param title Optional. Snackbar title
     */
    showSnackbar(message, type, title) {
        return MessageInvoker.postAndAwait(this.worker, SNACKBAR_SERVICE_MESSAGE_TYPE_BASE + 'showSnackbar', {
            message,
            type,
            title,
        });
    }
}

const SMART_SKILL_INSTANCE_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.SmartSkillInstanceService.';
/**
 * Service that handles updating smart skill instances
 */
class SmartSkillInstanceService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        this.eventTarget = eventTarget;
    }
    /**
     * Update the program node
     * @param instance New smart skill instance data
     */
    async updateInstance(instance) {
        return MessageInvoker.postAndAwait(this.eventTarget, SMART_SKILL_INSTANCE_SERVICE_MESSAGE_TYPE_BASE + 'updateInstance', {
            instance,
        });
    }
}

const PROGRAM_EXECUTION_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.ProgramExecutionService.';
/**
 * Service that handles program execution
 */
class ProgramExecutionService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(PROGRAM_EXECUTION_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Stop the program
     */
    async stopProgram() {
        return this.postAndAwait('stopProgram');
    }
    /**
     * Play the program
     */
    async playProgram() {
        return this.postAndAwait('playProgram');
    }
    /**
     * Resumes the program
     */
    async resumeProgram() {
        return this.postAndAwait('resumeProgram');
    }
    /**
     * Pause the program
     */
    async pauseProgram() {
        return this.postAndAwait('pauseProgram');
    }
    /**
     * Get the Program Status
     */
    getProgramStatus() {
        return this.postAndObserve('getStatus');
    }
    /**
     * Step to the next line in the program
     */
    async stepProgram() {
        return this.postAndAwait('stepProgram');
    }
}

const TOOL_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.ToolService.';
/**
 * Provides access to URScript functions
 * Parameters are aligned with URScript functions, see URScript manual for more info.
 */
class ToolService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(TOOL_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Sets the active tcp offset, i.e. the transformation from the output flange coordinate system to the TCP as a pose.
     * @see ScriptBuilder#setTcp
     */
    setTcp(tcp) {
        return this.postAndAwait('setTCP', tcp);
    }
    /**
     * Sets the mass, CoG (center of gravity), the inertia matrix of the active payload
     * @see ScriptBuilder#setTargetPayload
     */
    setTargetPayload(mass, cog, inertia) {
        return this.postAndAwait('setTargetPayload', { mass, cog, inertia });
    }
    /**
     * Get active TCP name
     * @returns `Observable<string>` emitting the name of the currently active TCP known to the controller.
     * If the TCP has been set without providing a name, this `Observable` will emit an empty `string`.
     * For documentation on `Observable`, see https://rxjs.dev/guide/observable
     */
    getActiveTCPName() {
        return this.postAndObserve('getActiveTCPName');
    }
}

const PROGRAM_CODE_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.ProgramCodeService.';
/**
 * Service that handles code generation
 */
class ProgramCodeService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        this.eventTarget = eventTarget;
    }
    /**
     *
     */
    async generateModulesScriptCode() {
        return MessageInvoker.postAndAwait(this.eventTarget, PROGRAM_CODE_SERVICE_MESSAGE_TYPE_BASE + 'GENERATE_MODULES_SCRIPT_CODE', {});
    }
    /**
     *
     */
    async generateGlobalFunctionsScriptCode() {
        return MessageInvoker.postAndAwait(this.eventTarget, PROGRAM_CODE_SERVICE_MESSAGE_TYPE_BASE + 'GENERATE_FUNCTIONS_SCRIPT_CODE', {});
    }
    /**
     *
     */
    async generateModulesAndGlobalFunctionsScriptCode() {
        return MessageInvoker.postAndAwait(this.eventTarget, PROGRAM_CODE_SERVICE_MESSAGE_TYPE_BASE + 'GENERATE_MODULES_AND_FUNCTIONS_SCRIPT_CODE', {});
    }
    /**
     *
     */
    async getProgramPreambleScriptCode() {
        return MessageInvoker.postAndAwait(this.eventTarget, PROGRAM_CODE_SERVICE_MESSAGE_TYPE_BASE + 'GET_PROGRAM_PREAMBLE_SCRIPT_CODE', {});
    }
}

const MOUNTING_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.MountingService.';
/**
 * Provides access to URScript functions
 * Parameters in this case are not aligned with setGravity
 * The input parameters are transformed into parameters for the setGravity URScript function
 * @see ScriptBuilder#setGravity
 */
class MountingService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(MOUNTING_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Sets the robot mounting as a rotation and a tilt
     * @param rotation The robot's rotation at the base
     * @param tilt The robot's tilt at the base
     */
    setMounting(rotation, tilt) {
        return this.postAndAwait('setMounting', { rotation, tilt });
    }
}

const FRAMES_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.FrameService.';
/**
 * Create, Read, Update, Delete functionality for frames
 * Parameters are aligned with URScript functions, see URScript manual for more info.
 */
class FramesService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(FRAMES_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Add a frame with the name 'name' initialized at the specified pose expressed in the refFrame coordinate frame.
     * This command only adds a frame to the world, it does not attach it to the ref_frame coordinate frame.
     * Use attach_frame() to attach the newly added frame to ref_frame if desired.
     * @see ScriptBuilder#addFrame
     */
    addFrame(name, pose, refFrame = 'base') {
        return this.postAndAwait('addFrame', { name, pose, refFrame });
    }
    /**
     * Delete the frame named frame from the world model.
     * The "world", "base", and "tcp" frames cannot be deleted.
     * Any frames that are attached to the deleted frame will be attached to the "world" frame with new frame offsets set
     * such that the detached frame does not move in the world.
     * This command will fail if the frame does not exist.
     * @see ScriptBuilder#deleteFrame
     */
    deleteFrame(name) {
        return this.postAndAwait('deleteFrame', { name });
    }
    /**
     * Changes the placement of the coordinate frame named name to the new placement given by pose that is defined in the refFrame coordinate frame.
     * This will fail if name is “world”, "tcp", or if the frame does not exist. Note: to move the "tcp" frame, use the set_tcp() command instead.
     * If being used with the part positioner product, the ref_name argument can be the name of an external axis or axis group.
     * @see ScriptBuilder#moveFrame
     */
    moveFrame(name, pose, refFrame = 'base') {
        return this.postAndAwait('moveFrame', { name, pose, refFrame });
    }
    /**
     * Changes the orientation of the coordinate frame named name to the new orientation.
     * This will fail if name is “world”, "tcp", or if the frame does not exist. Note: to reorient the "tcp" frame, use the set_tcp() command instead.
     * @see ScriptBuilder#setFrameOrientation
     */
    setFrameOrientation(name, orientation) {
        return this.postAndAwait('setFrameOrientation', { name, orientation });
    }
    /**
     * Attaches the child frame to the parent world model object. The relative transform between the parent and child will be set such that
     * the child does not move in the world when the attachment occurs.
     * The child cannot be "world", "tcp", or the same as parent.
     * This will fail if child or parent is not an existing frame, or this makes the attachments form a closed chain.
     * @see ScriptBuilder#attachFrame
     */
    attachFrame(child, parent) {
        return this.postAndAwait('attachFrame', { child, parent });
    }
    /**
     * Returns an observable of live frame data.
     */
    observeFrames() {
        return this.postAndObserve('observeFrames');
    }
}

const OPERATOR_SCREEN_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.OperatorScreenService.';
/**
 * Service for handling updating of operator screen instances
 */
class OperatorScreenService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(OPERATOR_SCREEN_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Update the operator screen instance
     * @param operatorScreen New operator screen instance data
     */
    async updateOperatorScreen(operatorScreen) {
        return this.postAndAwait('updateOperatorScreen', { operatorScreen });
    }
}

const ROBOT_INFO_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.RobotInfoService.';
/**
 * Provides access to information on the robot
 */
class RobotInfoService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(ROBOT_INFO_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Get the robot type
     */
    getRobotType() {
        return this.postAndAwait('getRobotType');
    }
    /**
     * Get the robot serial number
     */
    getSerialNumber() {
        return this.postAndAwait('getSerialNumber');
    }
    /**
     * Checks if the robot is in 3 PE stop state.
     *
     * @returns {Observable<boolean>} - An observable that emits `true` if the robot
     * is in 3 PE stop state, otherwise `false`.
     */
    is3PEStop() {
        return this.postAndObserve('is3PEStop');
    }
    /**
     * Determines whether the robot is currently running.
     *
     * @returns {Observable<boolean>} An observable that emits `true` if the robot is running
     * (`robotMode === RUNNING`), otherwise `false`.
     */
    isRobotRunning() {
        return this.postAndObserve('isRobotRunning');
    }
}

const FREEDRIVE_STATE_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.FreedriveStateService.';
/**
 * Provides access to freedrive information on the robot
 */
class FreedriveService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(FREEDRIVE_STATE_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Get the freedrive state
     *
     * @returns {Observable<boolean>} An observable that emits `true` if the robot is in freedrive state
     */
    isInFreedrive() {
        return this.postAndObserve('isInFreedrive');
    }
}

const PROGRAM_TREE_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.ProgramTreeService.';
/**
 * Service that handles node insertion and removal of nodes
 */
class ProgramTreeService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(PROGRAM_TREE_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Update a specific node in the program tree using the given nodes id
     * @param node Node to be updated
     */
    async updateSubTreeNode(node) {
        return this.postAndAwait('updateSubTreeNode', { node: node.node, id: node.id });
    }
    /**
     * Update child node
     * @param node Child node to be updated
     * @deprecated Use {@link updateSubTreeNode} instead
     */
    async updateChildNode(node) {
        return this.postAndAwait('updateChildNode', node);
    }
    /**
     * Add node as either child or sibling
     * @param node node to be added
     */
    async addNode(node) {
        return this.postAndAwait('addNode', node);
    }
    /**
     * Remove child node
     * @param node Child node to be removed
     */
    async removeChildNode(node) {
        return this.postAndAwait('removeChildNode', node);
    }
    /**
     * Sets the given node as the active program block
     * @param nodeID The node to navigate to
     */
    setActiveProgramBlock(nodeID) {
        return this.postAndAwait('setActiveProgramBlock', nodeID);
    }
    /**
     * Sets the given node as the active program block
     * @param nodeID The node to navigate to
     */
    selectProgramNode(nodeID) {
        return this.postAndAwait('selectProgramNode', nodeID);
    }
    /**
     * Get a readonly version of a given node
     * @param nodeID The node to get
     */
    getProgramNodeById(nodeID) {
        return this.postAndAwait('getProgramNodeById', nodeID);
    }
    /**
     * Get the tree context from a given program node
     */
    getTreeContextFromProgramNodeById(subTreeNodeId) {
        return this.postAndAwait('getTreeContextFromProgramNodeById', { subTreeNodeId });
    }
    /**
     * Get the contributed node instances for the URCap that are currently inserted in the program
     */
    getContributedNodeInstancesForURCap() {
        return this.postAndAwait('getContributedNodeInstancesForURCap');
    }
}

const SETTINGS_MESSAGE_TYPE_BASE = 'ur.web.message.SettingsService.';
/**
 *
 */
class SettingsService extends MessageInvokerService {
    /**
     *
     */
    constructor(eventTarget) {
        super(SETTINGS_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Get the current robot settings as an observable
     */
    getRobotSettings() {
        return this.postAndObserve('getRobotSettings');
    }
    /**
     * Get the current robot settings once
     */
    getRobotSettingsOnce() {
        return this.postAndAwait('getRobotSettingsOnce');
    }
}

const PROGRAM_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.ProgramService.';
/**
 * Service for handling program related operations
 */
class ProgramService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(PROGRAM_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Will get information and metadata of all programs
     * @returns Promise<Array<ProgramInformation>>
     */
    async getProgramsList() {
        return this.postAndAwait('getProgramsList');
    }
    /**
     * Get a program
     * @param programInformation
     * @returns Promise<Program>
     */
    async getProgram(programInformation) {
        return this.postAndAwait('getProgram', { programInformation });
    }
    /**
     * Load program
     * @param programId The id of the program to load
     * @see {@link getPrograms} for getting the ids
     */
    loadProgram({ programId }) {
        return this.postAndAwait('loadProgram', { programId });
    }
    /**
     * Retrieves the loaded program's information.
     *
     * This method sends a request to the backend to retrieve the information of the currently loaded program.
     *
     * @returns {Promise<ProgramInformation>} A promise that resolves to the information of the loaded program.
     *
     * @example
     * // Usage example
     * const programInfo = await this.presenterAPI.programService.getLoadedProgram()
     * console.log('Loaded Program Information:', programInfo);
     *
     */
    getLoadedProgram() {
        return this.postAndAwait('getLoadedProgram');
    }
}

const SYSTEM_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.SystemService.';
/**
 * Service that handles system information.
 */
class SystemService {
    /**
     * Create a service for the specific worker
     * @param worker
     */
    constructor(worker) {
        this.worker = worker;
    }
    /**
     * Get the software version.
     * @returns The software version.
     */
    getSoftwareVersion() {
        return MessageInvoker.postAndAwait(this.worker, SYSTEM_SERVICE_MESSAGE_TYPE_BASE + 'getSoftwareVersion');
    }
}

const MOTION_PROFILES_SERVICE_MESSAGE_TYPE_BASE = 'ur.web.message.MotionProfilesService.';
/**
 * Service for getting information of motion profiles
 */
class MotionProfilesService extends MessageInvokerService {
    /**
     * @constructor
     * @param eventTarget
     */
    constructor(eventTarget) {
        super(MOTION_PROFILES_SERVICE_MESSAGE_TYPE_BASE, eventTarget);
        this.eventTarget = eventTarget;
    }
    /**
     * Returns the data stored on the specified Motion Profiles type
     * @param type the type of the Motion Profiles to fetch:
     * - joint
     * - linear
     * - process
     */
    getMotionProfiles(type) {
        return this.postAndAwait('getMotionProfiles', { type });
    }
}

/**
 * Enum for HistoryPolicy
 * @readonly
 * @enum {number}
 */
var HistoryPolicy;
(function (HistoryPolicy) {
    HistoryPolicy[HistoryPolicy["SYSTEM_DEFAULT"] = 0] = "SYSTEM_DEFAULT";
    HistoryPolicy[HistoryPolicy["KEEP_LAST"] = 1] = "KEEP_LAST";
    HistoryPolicy[HistoryPolicy["KEEP_ALL"] = 2] = "KEEP_ALL";
    HistoryPolicy[HistoryPolicy["UNKNOWN"] = 3] = "UNKNOWN";
})(HistoryPolicy || (HistoryPolicy = {}));
/**
 * Enum for ReliabilityPolicy
 * @readonly
 * @enum {number}
 */
var ReliabilityPolicy;
(function (ReliabilityPolicy) {
    ReliabilityPolicy[ReliabilityPolicy["SYSTEM_DEFAULT"] = 0] = "SYSTEM_DEFAULT";
    ReliabilityPolicy[ReliabilityPolicy["RELIABLE"] = 1] = "RELIABLE";
    ReliabilityPolicy[ReliabilityPolicy["BEST_EFFORT"] = 2] = "BEST_EFFORT";
    ReliabilityPolicy[ReliabilityPolicy["UNKNOWN"] = 3] = "UNKNOWN";
})(ReliabilityPolicy || (ReliabilityPolicy = {}));
/**
 * Enum for DurabilityPolicy
 * @readonly
 * @enum {number}
 */
var DurabilityPolicy;
(function (DurabilityPolicy) {
    DurabilityPolicy[DurabilityPolicy["SYSTEM_DEFAULT"] = 0] = "SYSTEM_DEFAULT";
    DurabilityPolicy[DurabilityPolicy["TRANSIENT_LOCAL"] = 1] = "TRANSIENT_LOCAL";
    DurabilityPolicy[DurabilityPolicy["VOLATILE"] = 2] = "VOLATILE";
    DurabilityPolicy[DurabilityPolicy["UNKNOWN"] = 3] = "UNKNOWN";
})(DurabilityPolicy || (DurabilityPolicy = {}));
/**
 * QoS liveliness enumerations that describe a publisher's reporting policy for its alive status.
 * For a subscriber, these are its requirements for its topic's publishers.
 * @readonly
 * @enum {number}
 */
var LivelinessPolicy;
(function (LivelinessPolicy) {
    LivelinessPolicy[LivelinessPolicy["SYSTEM_DEFAULT"] = 0] = "SYSTEM_DEFAULT";
    LivelinessPolicy[LivelinessPolicy["AUTOMATIC"] = 1] = "AUTOMATIC";
    LivelinessPolicy[LivelinessPolicy["MANUAL_BY_NODE"] = 2] = "MANUAL_BY_NODE";
    LivelinessPolicy[LivelinessPolicy["MANUAL_BY_TOPIC"] = 3] = "MANUAL_BY_TOPIC";
    LivelinessPolicy[LivelinessPolicy["UNKNOWN"] = 4] = "UNKNOWN";
})(LivelinessPolicy || (LivelinessPolicy = {}));
/**
 * The quality of service profile for ROS2 communication.
 * @dynamic
 */
class QoSProfile {
    /**
     * Get HistoryPolicy enum.
     * @return HistoryPolicy
     */
    static { this.HistoryPolicy = HistoryPolicy; }
    /**
     * Get ReliabilityPolicy enum.
     * @return ReliabilityPolicy
     */
    static { this.ReliabilityPolicy = ReliabilityPolicy; }
    /**
     * Get DurabilityPolicy enum.
     * @return DurabilityPolicy
     */
    static { this.DurabilityPolicy = DurabilityPolicy; }
    /**
     * Get LivelinessPolicy enum
     * @return LivelinessPolicy
     */
    static { this.LivelinessPolicy = LivelinessPolicy; }
    /**
     * Default duration is duration unspecified
     */
    static get DurationUnspecified() {
        return { seconds: 0, nanoseconds: 0 };
    }
    /**
     * Default duration is duration unspecified
     */
    static get DurationInfinite() {
        return { seconds: 9223372036, nanoseconds: 854775807 };
    }
    /**
     * Get default profile. Default values are:
     *     HistoryPolicy = KEEP_LAST,
     *     depth = 10,
     *     ReliabilityPolicy = RELIABLE,
     *     DurabilityPolicy = VOLATILE,
     *     deadline = QoSProfile.DurationUnspecified,
     *     lifespan = QoSProfile.DurationUnspecified,
     *     LivelinessPolicy = SYSTEM_DEFAULT,
     *     liveliness_lease_duration = QoSProfile.DurationUnspecified,
     *     avoid_ros_namespace_conventions = false
     * @return QoSProfile
     */
    static get default() {
        return new QoSProfile(HistoryPolicy.KEEP_LAST, 10, ReliabilityPolicy.RELIABLE, DurabilityPolicy.VOLATILE, QoSProfile.DurationUnspecified, QoSProfile.DurationUnspecified, LivelinessPolicy.SYSTEM_DEFAULT, QoSProfile.DurationUnspecified, false);
    }
    /**
     * Get most compatible QoS profile with Reliability set to BEST_EFFORT. Values are:
     *     HistoryPolicy = KEEP_LAST,
     *     depth = 10,
     *     ReliabilityPolicy = BEST_EFFORT,
     *     DurabilityPolicy = VOLATILE,
     *     deadline = QoSProfile.DurationUnspecified,
     *     lifespan = QoSProfile.DurationUnspecified,
     *     LivelinessPolicy = SYSTEM_DEFAULT,
     *     liveliness_lease_duration = QoSProfile.DurationUnspecified,
     *     avoid_ros_namespace_conventions = false
     * @return QoSProfile
     */
    static get compatible() {
        return new QoSProfile(HistoryPolicy.KEEP_LAST, 10, ReliabilityPolicy.BEST_EFFORT, DurabilityPolicy.VOLATILE, QoSProfile.DurationUnspecified, QoSProfile.DurationUnspecified, LivelinessPolicy.SYSTEM_DEFAULT, QoSProfile.DurationUnspecified, false);
    }
    /**
     * Create a QosProfile for ROS2 communication.
     * @param history HistoryPolicy The history value, default value is HistoryPolicy.KEEP_LAST.
     * @param depth number The depth value, default value is depth=10.
     * @param reliability ReliabilityPolicy The reliability value, default value is ReliabilityPolicy.RELIABLE.
     * @param durability DurabilityPolicy The durability value, default value is DurabilityPolicy.VOLATILE.
     * @param deadline The deadline value, default value is DurationUnspecified.
     * @param lifespan The lifespan value, default value is DurationUnspecified.
     * @param liveliness LivelinessPolicy The liveliness value, default value is LivelinessPolicy.SYSTEM_DEFAULT.
     * @param liveliness_lease_duration The liveliness_lease_duration value, default values is DurationUnspecified.
     * @param avoid_ros_namespace_conventions The avoid_ros_namespace_conventions value, default value=false.
     */
    constructor(history = HistoryPolicy.KEEP_LAST, depth = 10, reliability = ReliabilityPolicy.RELIABLE, durability = DurabilityPolicy.VOLATILE, deadline = QoSProfile.DurationUnspecified, lifespan = QoSProfile.DurationUnspecified, liveliness = LivelinessPolicy.SYSTEM_DEFAULT, liveliness_lease_duration = QoSProfile.DurationUnspecified, avoid_ros_namespace_conventions = false) {
        this.history = history;
        this.depth = depth;
        this.reliability = reliability;
        this.durability = durability;
        this.deadline = deadline;
        this.lifespan = lifespan;
        this.liveliness = liveliness;
        this.liveliness_lease_duration = liveliness_lease_duration;
        this.avoid_ros_namespace_conventions = avoid_ros_namespace_conventions;
    }
}

/**
 * Check if message is an exception, throw error if exception
 * @param msg message to check
 */
const assertNotException = (msg) => {
    if (isException(msg)) {
        throw new Error(msg.exception);
    }
};
/**
 * Check if message is an exception
 * @param msg
 * @private
 * @return boolean true if exception
 */
function isException(msg) {
    return msg.exception !== undefined && msg.message !== undefined;
}
/**
 * The UR TypeScript API for ROS 2
 */
class Ros2Client {
    /**
     * Get Id for connection
     * @return Id for connection
     * @private
     */
    getConnectionId() {
        return Guid.create().toString();
    }
    /**
     * Create Ros2Client ROS2 client for UR ROS2 Web Bridge
     * @constructor
     * @param ros2Service Ros2 service for ROS2 communication
     */
    constructor(ros2Service) {
        this.ros2Service = ros2Service;
    }
    /**
     * Get topic path to be used for ROS2 communication, including optional robot namespace and custom namespace.
     * @param prependRobotNamespace boolean if true then local ROS2 robot namespace will be prepended to topic.
     * @param topic string to be used for ROS2 communication.
     * @param namespace optional namespaces to prepend topic.
     * @return TopicPath extending Promise<string>
     */
    async getTopicPath(prependRobotNamespace, topic, ...namespace) {
        return await this.getPath(prependRobotNamespace, topic, namespace);
    }
    /**
     * Get service path to be used for ROS2 communication, including optional robot namespace and custom namespace.
     * @param prependRobotNamespace boolean if true then local ROS2 robot namespace will be prepended to service name.
     * @param serviceName string to be used for ROS2 communication.
     * @param namespace optional namespaces to prepend service name.
     * @return ServicePath extending Promise<string>
     */
    async getServicePath(prependRobotNamespace, serviceName, ...namespace) {
        return await this.getPath(prependRobotNamespace, serviceName, namespace);
    }
    /**
     * Get promise path
     * @param prependRobotNamespace
     * @param topicOrMessageName
     * @param namespace
     * @return Promise<String>
     * @private
     */
    async getPath(prependRobotNamespace, topicOrMessageName, namespace) {
        let localNamespace = namespace.length > 0 ? namespace.join('/').replace(/^\/?/, '').replace(/\/?$/, '') : '';
        localNamespace = localNamespace.length > 0 ? `/${localNamespace}/` : '';
        if (prependRobotNamespace) {
            const ros2Namespace = await this.getRobotNamespace();
            let rosNamespacePath = ros2Namespace?.length > 0 ? ros2Namespace.replace(/^\/?/, '/').replace(/\/?$/, '') : '';
            rosNamespacePath = rosNamespacePath.length > 0 && localNamespace.length === 0 ? rosNamespacePath + '/' : rosNamespacePath;
            return rosNamespacePath + localNamespace + topicOrMessageName;
        }
        return localNamespace + topicOrMessageName;
    }
    /**
     * Get RobotNamespace
     * @private
     */
    async getRobotNamespace() {
        if (!this.robotNamespace) {
            this.robotNamespace = await this.ros2Service.getRos2Namespace();
        }
        return this.robotNamespace;
    }
    /**
     * Create a new subscription to a ROS2 topic.
     * @param messageType The type of ROS messages the subscription will subscribe to.
     * @param topicPath TopicPath The name of the topic the subscription will subscribe to, may include ros2 namespace.
     * @param qosProfile optional quality of service profile to apply to the subscription. Default is QoSProfile.default.
     * @return Observable to subscribe to
     */
    createSubscription(messageType, topicPath, qosProfile = QoSProfile.default) {
        const connectionId = this.getConnectionId();
        return this.ros2Service.subscribe(connectionId, messageType, topicPath, qosProfile).pipe(tap(assertNotException), map$1((rt) => rt.msg));
    }
    /**
     * Create a new subscription for frame transformation
     * @param targetFrame
     * @param sourceFrame
     * @return Observable to subscribe to
     */
    createTransformSubscription(targetFrame, sourceFrame) {
        const connectionId = this.getConnectionId();
        return this.ros2Service.subscribeTransform(connectionId, targetFrame, sourceFrame).pipe(tap(assertNotException), switchMap((rt) => {
            if (rt.msg?.error) {
                return throwError(() => new Error(rt.msg.error));
            }
            return of(rt);
        }), map$1((rt) => rt.msg));
    }
    /**
     * Make a ROS2 service call for setting argumentValues on a service, and return an observable of type T.
     * @param serviceType The type of ROS service to set argumentValues on.
     * @param servicePath ServicePath The name of the service to set argumentValues on, may include namespace.
     * @param argumentValues Values to be set.
     * @param qosProfile optional quality of service profile to apply to the service client. Default is QoSProfile.default.
     * @param timeout optional float value in seconds, default is 30.0 seconds before timeout.
     * @return The first message returned from the service call.
     */
    async callService(serviceType, servicePath, argumentValues, qosProfile = QoSProfile.default, timeout = 30.0) {
        const connectionId = this.getConnectionId();
        return await firstValueFrom(this.ros2Service.callService(connectionId, serviceType, servicePath, argumentValues, qosProfile, timeout).pipe(tap(assertNotException), map$1((rt) => rt.msg)));
    }
}

const ROS2_MESSAGE_TYPE_BASE = 'ur.web.message.Ros2Service.';
/**
 * Service for ROS2 communication
 * Uses event based messaging for communication
 */
class Ros2EventService {
    /**
     * Create event based messaging implementation of the Ros2ServiceApi for ROS2 communication
     * @param target Either a Worker to be used as the EventTarget or a standard
     */
    constructor(target) {
        this.target = target;
    }
    /**
     * Get local ROS2 robot namespace
     * @return Promise containing local ROS2 robot namespace
     */
    getRos2Namespace() {
        return MessageInvoker.postAndAwait(this.target, ROS2_MESSAGE_TYPE_BASE + 'getRos2Namespace', {});
    }
    /**
     * Subscribe to topicPath with given message type using QoS profile
     * @param connectionId Id for subscription
     * @param messageType
     * @param topic
     * @param qosProfile optional quality of service profile to apply to the subscription. Default is QoSProfile.default.
     * @return Observable to subscribe to
     */
    subscribe(connectionId, messageType, topic, qosProfile = QoSProfile.default) {
        return MessageInvoker.postAndObserve(this.target, ROS2_MESSAGE_TYPE_BASE + 'subscribe', {
            connectionId,
            messageType,
            topic,
            qosProfile,
        });
    }
    /**
     * Subscribe to frame transformation
     * @param connectionId Id for subscription
     * @param targetFrame
     * @param sourceFrame
     * @param publishRate in milliseconds
     * @return Observable to subscribe to
     */
    subscribeTransform(connectionId, targetFrame, sourceFrame, publishRate = 100) {
        return MessageInvoker.postAndObserve(this.target, ROS2_MESSAGE_TYPE_BASE + 'subscribeTransform', {
            connectionId,
            targetFrame,
            sourceFrame,
            publishRate,
        });
    }
    /**
     * Call service to set argumentValues of service type on service name
     * @param connectionId Id for subscription
     * @param serviceType
     * @param serviceName
     * @param argumentValues
     * @param qosProfile optional quality of service profile to apply to the service client. Default is QoSProfile.default.
     * @param timeout optional float value in seconds, default is 30.0 seconds before timeout.
     * @return Promise with value
     */
    callService(connectionId, serviceType, serviceName, argumentValues, qosProfile = QoSProfile.default, timeout = 30.0) {
        return MessageInvoker.postAndObserve(this.target, ROS2_MESSAGE_TYPE_BASE + 'callService', {
            connectionId,
            serviceType,
            serviceName,
            argumentValues,
            qosProfile,
            timeout,
        });
    }
}

/**
 * Provides access to common related API features in both BehaviorWorkers
 * and Presenters
 */
class CommonBehaviorAPI {
    /** Creates a new instance of the CommonAPI based on either a Worker
     * or an instance of an EventTarget, to allow communication with the
     * surrounding apis from within the Worker or WebComponent
     *
     * @param target Either a Worker to be used as the EventTarget or a standard
     * EventTarget
     */
    constructor(target) {
        this.applicationService = new ApplicationService(target);
        this.logMessageService = new LogMessageService(target);
        this.logPerformanceService = new LogPerformanceService(target);
        this.ros2Client = new Ros2Client(new Ros2EventService(target));
        this.safetyService = new SafetyService(target);
        this.sourceService = new SourceService(target);
        this.robotPositionService = new RobotPositionService(target);
        this.symbolService = new SymbolService(target);
        this.robotInfoService = new RobotInfoService(target);
        this.programTreeService = new ProgramTreeService(target);
        this.settingsService = new SettingsService(target);
        this.systemService = new SystemService(target);
    }
    /**
     * Creates a URL string for communicating with backend Container Contributions
     * A protocol must be prepended to the string, like 'http://' or 'ws://'
     * @param vendorID Id of Vendor as specified in manifest.yaml file
     * @param urcapID Id of URCap as specified in manifest.yaml file
     * @param containerName Name of Container as specified in manifest.yaml file
     * @param serviceName Name of Service as specified by Ingress name in manifest.yaml file
     * @return URL string without protocol.
     */
    getContainerContributionURL(vendorID, urcapID, containerName, serviceName) {
        return `${location.host}/${vendorID}/${urcapID}/${containerName}/${serviceName}`;
    }
}
/**
 * Provides access to common related API features in Presenters
 */
class CommonPresenterAPI extends CommonBehaviorAPI {
    /** Creates a new instance of the CommonAPI based on an instance of an EventTarget,
     * to allow communication with the
     * surrounding apis from within the Worker or WebComponent
     *
     * @param target An EventTarget
     */
    constructor(target) {
        super(target);
        this.dialogService = new DialogService(target);
        this.snackbarService = new SnackbarService(target);
        this.programService = new ProgramService(target);
    }
}

/**
 * Provides access to Application related API features in ApplicationBehaviors
 */
class ApplicationBehaviorAPI extends CommonBehaviorAPI {
    /** Creates a new instance of the ApplicationBehaviorAPI to allow communication with the
     * surrounding application from within a WebWorker
     *
     * @param target A worker, provided with 'self' when created inside a web worker
     */
    constructor(target) {
        super(target);
        this.toolService = new ToolService(target);
        this.mountingService = new MountingService(target);
        this.framesService = new FramesService(target);
    }
}
/**
 * Provides access to Application related API features in ApplicationPresenters
 */
class ApplicationPresenterAPI extends CommonPresenterAPI {
    /** Creates a new instance of the ApplicationPresenterAPI, to allow communication with the
     * surrounding application from within a WebComponent
     *
     * @param target An EventTarget, provided by PolyScope X
     */
    constructor(target) {
        super(target);
        this.applicationNodeService = new ApplicationNodeService(target);
        this.robotMoveService = new RobotMoveService(target);
        this.programExecutionService = new ProgramExecutionService(target);
        this.toolService = new ToolService(target);
        this.mountingService = new MountingService(target);
        this.framesService = new FramesService(target);
    }
}

/**
 * Provides access to Program related API features in both BehaviorWorkers
 * and ProgramPresenters
 */
class ProgramBehaviorAPI extends CommonBehaviorAPI {
    /** Creates a new instance of the ProgramBehaviorAPI to allow communication with the
     * surrounding application from within a WebWorker
     *
     * @param target A worker, provided with 'self' when created inside a web worker
     */
    constructor(target) {
        super(target);
        this.builder = new TreeBuilder(target);
        this.framesService = new FramesService(target);
        this.motionProfilesService = new MotionProfilesService(target);
    }
}
/**
 * ProgramPresenterAPI
 */
class ProgramPresenterAPI extends CommonPresenterAPI {
    /** Creates a new instance of the ProgramPresenterAPI, to allow communication with the
     * surrounding application from within a WebComponent
     *
     * @param target An EventTarget, provided by PolyScope X
     * @param selectedNodeId the id of the currently selected program node
     */
    constructor(target, selectedNodeId) {
        super(target);
        this.selectedNodeId = selectedNodeId;
        this.programNodeService = new ProgramNodeService(target, selectedNodeId);
        this.robotMoveService = new RobotMoveService(target);
        this.validationService = new ValidationService(target, selectedNodeId);
        this.builder = new TreeBuilder(target);
        this.framesService = new FramesService(target);
        this.motionProfilesService = new MotionProfilesService(target);
    }
}

/**
 * Provides access to source related API features in Source Behaviors
 */
class SourceBehaviorAPI extends CommonBehaviorAPI {
    /** Creates a new instance of the SourceBehaviorAPI to allow communication with the
     * surrounding application from within a WebWorker
     *
     * @param target A worker, provided with 'self' when created inside a web worker
     */
    constructor(target) {
        super(target);
    }
}
/**
 * Provides access to source related API features in Source Presenters
 */
class SourcePresenterAPI extends CommonPresenterAPI {
    /** Creates a new instance of the SourcePresenterAPI, to allow communication with the
     * surrounding application from within a WebComponent
     *
     * @param target An EventTarget, provided by PolyScope X
     * @param sourceID ID for this source
     */
    constructor(target, sourceID) {
        super(target);
        this.sourceID = sourceID;
        this.signalService = new SignalService(target, sourceID);
        this.sourceNodeService = new SourceNodeService(target);
    }
}

/**
 * Provides access to Smart Skills related API features in BehaviorWorkers
 */
class SmartSkillsBehaviorAPI extends CommonBehaviorAPI {
    /** Creates a new instance of the SmartSkillsBehaviorAPI to allow communication with the
     * surrounding application from within a WebWorker
     *
     * @param target A worker, provided with 'self' when created inside a web worker
     */
    constructor(target) {
        super(target);
        this.builder = new TreeBuilder(target);
        this.programCode = new ProgramCodeService(target);
    }
}
/**
 *
 */
class SmartSkillsPresenterAPI extends CommonPresenterAPI {
    /**
     * Create a new instance of the SmartSkillsPresenterAPI to allow a Presenter to interact with the API and update itself
     */
    constructor(target, name) {
        super(target);
        this.smartSkillInstanceService = new SmartSkillInstanceService(target);
        this.robotMoveService = new RobotMoveService(target);
        this.freedriveService = new FreedriveService(target);
    }
}

/**
 * Checks that the message coming is in a SerializableApplicationContext
 * @param message The message to serialize
 */
function isSerializedApplicationContext(message) {
    return message && message.type === '$$ApplicationContext';
}
/**
 * Creates the serializer for the ApplicationContext. Handles serializing and deserializing of the object.
 * @constructor
 */
const ApplicationContextSerializer = () => {
    return {
        /**
         * deserialize the message using a given handler
         * @param message
         * @param defaultHandler
         */
        deserialize(message, defaultHandler) {
            if (isSerializedApplicationContext(message)) {
                // created the application context from the message
                return {
                    contributions: {
                        contributionList: message.contributions?.contributionList ?? [],
                        /**
                         * Returns a Contribution node base on the param
                         * @param type String specifying the contribution node to fetch
                         */
                        getContribution(type) {
                            return this.contributionList.find((contribution) => contribution.parentType === type);
                        },
                    },
                    frames: {
                        framesList: message.frames?.framesList ?? [],
                        /**
                         * Returns a Frame base on the param
                         * @param frameId String specifying the frame to fetch
                         */
                        getFrame(frameId) {
                            return this.framesList.find((frame) => frame.name === frameId);
                        },
                    },
                };
            }
            else {
                return defaultHandler(message);
            }
        },
        /**
         * serializing the item using a give handler
         * @param item
         * @param defaultHandler
         */
        serialize(item, defaultHandler) {
            if (isSerializedApplicationContext(item)) {
                return {
                    type: '$$ApplicationContext',
                    contributions: {
                        contributionList: item.contributions?.contributionList ?? [],
                    },
                    frames: {
                        framesList: item.frames?.framesList ?? [],
                    },
                };
            }
            else {
                return defaultHandler(item);
            }
        },
    };
};

/**
 * An asynchronous array iterator
 */
class AsyncArrayTraverser {
    /**
     * @param nodes the array of nodes to iterate
     */
    constructor(nodes) {
        this.pointer = 0;
        this.values = nodes;
    }
    /**
     * provide the next element for the iterator
     */
    async next() {
        if (this.pointer < this.values.length) {
            const index = this.pointer;
            this.pointer++;
            return {
                done: false,
                value: this.values[index],
            };
        }
        return {
            done: true,
            value: null,
        };
    }
    /**
     * returns the asynchronous iterator
     */
    [Symbol.asyncIterator]() {
        return this;
    }
}

/**
 * Checks that the message coming is in a isSerializedInsertionContext
 * @param message
 */
function isSerializedInsertionContext(message) {
    return message && message.type === '$$InsertionContext';
}
/**
 * Creates the serializer for the InsertionContext. Handles serializing and deserializing of the object.
 * @constructor
 */
const InsertionContextSerializer = () => {
    return {
        /**
         * Deserializes the message as a new InsertionContext based on the message inputs
         * @param message Message containing the values to be parsed to the new InsertionContext
         * @param defaultHandler The object to handle deserializing if not of type InsertionContext
         */
        deserialize(message, defaultHandler) {
            if (isSerializedInsertionContext(message)) {
                return {
                    traverse: {
                        ancestors: new AsyncArrayTraverser(message.ancestors ? message.ancestors : []),
                        nextSiblings: new AsyncArrayTraverser(message.nextSiblings ? message.nextSiblings : []),
                        previousSiblings: new AsyncArrayTraverser(message.previousSiblings ? message.previousSiblings : []),
                    },
                };
            }
            else {
                return defaultHandler(message);
            }
        },
        /**
         * Serializes the object and inserts a type to be used when deserializing to proper identifying the object when returning
         * @param item The original item that will be serialized
         * @param defaultHandler The object to handle serializing if not of type InsertionContext
         */
        serialize(item, defaultHandler) {
            if (isSerializedInsertionContext(item)) {
                return {
                    type: '$$InsertionContext',
                    ancestors: item.ancestors,
                    previousSiblings: item.previousSiblings,
                    nextSiblings: item.nextSiblings,
                };
            }
            else {
                return defaultHandler(item);
            }
        },
    };
};

/**
 *
 * ScriptBuilder provides methods to easily generate well indented UR script programmatically.
 * This class manages indentation automatically as the script is built step by step.
 * But,it does not guarantee perfect formatting as there is a possibility to pass badly formatted expressions, statements or scripts.
 *
 */
class ScriptBuilder {
    static { this.SINGLE_INDENT = '  '; }
    static { this.SINGLE_INDENT_LEVEL = 1; }
    /**
     * Creates a new ScriptBuilder
     *
     * @param script If a script is specified, the new ScriptBuilder will be preloaded with that script
     * @param initialIndent If an indent is specified, the ScriptBuilder will be initialized with that indent
     */
    constructor(script = '', initialIndent = 0) {
        this.type = '$$ScriptBuilder';
        this.script = '';
        this.currentIndent = 0;
        this.addStatements(script);
        this.currentIndent = initialIndent;
    }
    /**
     * Generate the String representation of a boolean.
     *
     * @param b The input boolean
     * @return The String representation of the input
     */
    static getBooleanAsString(b) {
        return b ? 'True' : 'False';
    }
    /**
     * Return a String representing the URScript for negating an expression.
     * @param expression The expression to be negated
     * @return String representation of the negated expression.
     */
    /**
     *
     * @param expression The expression to be negated
     */
    static negate(expression) {
        return `not (${expression})`;
    }
    /**
     * Return a String representation of the input array.
     * @param input the input array.
     * @return String representation of the input array.
     */
    static formatArray(input) {
        let stringElem;
        let result;
        result = '[';
        input.forEach((elem, index) => {
            if (typeof elem === 'boolean') {
                stringElem = ScriptBuilder.getBooleanAsString(elem);
            }
            else if (typeof elem === 'string') {
                stringElem = `"${elem}"`;
            }
            else {
                stringElem = elem.toString();
            }
            result += stringElem;
            if (index < input.length - 1) {
                result += ', ';
            }
        });
        return result + ']';
    }
    /**
     * Create a ScriptBuilder preloaded with the incognito program definition.
     * This is unbalanced, it has to be balanced by a call to {@link #end()} before
     * calling {@link #getScript()}.
     *
     * @return new ScriptBuilder.
     */
    static CreateInkognitoProgram() {
        return new ScriptBuilder().defineFunction('inkognito');
    }
    /**
     * Create a ScriptBuilder preloaded with a secondary program definition.
     * This is unbalanced, it has to be balanced by a call to {@link #end()} before
     * calling {@link #getScript()}.
     *
     * @param name The name of the secondary program.
     * @return new ScriptBuilder.
     */
    static CreateSecondaryProgram(name) {
        const returnValue = new ScriptBuilder();
        returnValue.addStatements(`sec ${name}():`);
        returnValue.increaseIndent();
        return returnValue;
    }
    /**
     * Add a block of statements to the builder.
     *
     * @param statements A block of statements as a String.
     * @return this object.
     */
    addStatements(statements) {
        // Convert indention to a string
        // If currentIndent is negative, then use blank string
        let indentString = '';
        for (let i = 0; i < this.currentIndent; i++) {
            indentString += ScriptBuilder.SINGLE_INDENT;
        }
        statements
            .split('\n')
            .filter((statement) => statement.length > 0)
            .forEach((statement) => {
            this.script += `${indentString}${statement}\n`;
        });
        return this;
    }
    /**
     * Add raw script code.
     * As this method takes raw script code that may not be indented and appends it without indenting it,
     * it may violate the overall indentation.
     *
     * @param script The script to be appended
     * @return this object.
     */
    addRaw(script) {
        this.script += `${script}\n`;
        return this;
    }
    /**
     * Add a while statement with a given expression.
     * This is unbalanced, it has to be balanced by a call to {@link #end()} at the appropriate time.
     *
     * @param expression The string representation of an expression.
     * @return this object.
     */
    beginWhile(expression) {
        this.addStatements(`while (${expression}):`);
        this.increaseIndent();
        return this;
    }
    /**
     * Add a while not statement with a given expression.
     * This is unbalanced, it has to be balanced by a call to {@link #end()} at the appropriate time.
     *
     * @param expression The string representation of an expression.
     * @return this object.
     */
    beginWhileNot(expression) {
        this.addStatements(`while not (${expression}):`);
        this.increaseIndent();
        return this;
    }
    /**
     * Add a while statement that loops forever.
     * This is unbalanced, it has to be balanced by a call to {@link #end()} at the appropriate time.
     *
     * @return this object.
     */
    beginWhileTrue() {
        this.beginWhile('True');
        return this;
    }
    /**
     * Add the end keyword that balances/terminates the declaration of functions, branch and loop statements.
     *
     * @return this object.
     */
    end() {
        this.decreaseIndent();
        this.addStatements('end');
        return this;
    }
    /**
     * End a block, without an end statement
     * Used by if and elseif
     * @return this object.
     */
    endBlock() {
        this.decreaseIndent();
        return this;
    }
    /**
     * Add a return statement.
     *
     * @return this object.
     */
    returnMethod() {
        this.addStatements('return');
        return this;
    }
    /**
     * Add a statement that declares a global variable and assigns it an initial value.
     *
     * @param name  The name of the global variable.
     * @param value An expression which is the initial value of the global variable.
     * @return this object.
     */
    globalVariable(name, value) {
        this.addStatements(`global ${name} = ${value}`);
        return this;
    }
    /**
     * Add a statement that declares a local variable and assigns it an initial value.
     *
     * @param name  The name of the local variable.
     * @param value An expression which is the initial value of the local variable.
     * @return this object.
     */
    localVariable(name, value) {
        this.addStatements(`local ${name} = ${value}`);
        return this;
    }
    /**
     * Add a statement that increments a variable with the given name.
     *
     * @param name The name of the variable.
     * @return this object.
     */
    incrementVariable(name) {
        this.addStatements(`${name} = ${name} + 1`);
        return this;
    }
    /**
     * Add a statement that assigns an expression to a variable.
     *
     * @param name       The name of the variable.
     * @param expression The expression to be assigned to the variable.
     * @return this object.
     */
    assign(name, expression) {
        this.addStatements(`${name} = ${expression}`);
        return this;
    }
    /**
     * Add an if statement with a given expression.
     * This is unbalanced, it has to be balanced by a call to {@link #elseIfCondition(String)} or
     * {@link #else()} or {@link #end()} at the appropriate time.
     *
     * @param expression The expression that the if statement branches on.
     * @return this object.
     */
    ifCondition(expression) {
        this.addStatements(`if (${expression}):`);
        this.increaseIndent();
        return this;
    }
    /**
     * Add an else if statement with a given expression.
     * This is unbalanced, it has to be balanced by a call to {@link #elseIfCondition(String)} or
     * {@link #elseIfCondition(Script)} or {@link #else()} or {@link #end()} at the appropriate time.
     *
     * @param expression The expression that the else if statement branches on.
     * @return this object.
     */
    elseIfCondition(expression) {
        this.addStatements(`elif (${expression}):`);
        this.increaseIndent();
        return this;
    }
    /**
     * Add an else statement.
     * This is unbalanced, it has to be balanced by a call to {@link #end()} at the appropriate time.
     *
     * @return this object.
     */
    else() {
        this.addStatements('else:');
        this.increaseIndent();
        return this;
    }
    /**
     * Add a thread definition with given the threadName.
     * This is unbalanced, it has to be balanced by a call to {@link #end()} at the appropriate time.
     *
     * @param threadName The threadName of the thread.
     * @return this object.
     */
    defineThread(threadName) {
        this.addStatements(`thread ${threadName}():`);
        this.increaseIndent();
        return this;
    }
    /**
     * Add a statement to run a thread with the given name and assign the thread handle to a variable.
     *
     * @param threadHandle The variable that gets assigned the thread handle .
     * @param threadName   The name of the thread to be started.
     * @return this object.
     */
    runThread(threadHandle, threadName) {
        this.addStatements(`${threadHandle} = run ${threadName}()`);
        return this;
    }
    /**
     * Add a statement to kill a thread.
     *
     * @param threadHandle The thread threadHandle.
     * @return this object.
     */
    killThread(threadHandle) {
        this.addStatements(`kill ${threadHandle}`);
        return this;
    }
    /**
     * Add a function definition with the given name and parameters.
     * This is unbalanced, it has to be balanced by a call to {@link #end()} at the appropriate time.
     *
     * @param name   The name of the function.
     * @param params The parameters of the function.
     * @return this object.
     */
    defineFunction(name, ...params) {
        this.addStatements(`def ${name}(${params.join(',')}):`);
        this.increaseIndent();
        return this;
    }
    /**
     * Add a comment in the script.
     *
     * @param text The comment text.
     * @return this object.
     */
    comment(text) {
        for (const comment of text.split('\n')) {
            this.addStatements(`# ${comment}`);
        }
        return this;
    }
    /**
     * Add a frame with the name 'name' initialized at the specified pose expressed in the refFrame coordinate frame.
     * This command only adds a frame to the world, it does not attach it to the ref_frame coordinate frame.
     * Use attach_frame() to attach the newly added frame to ref_frame if desired.
     * @param name name of the frame to be added.  The name must not be the same as any existing world model object (frame, axis, or axis group), otherwise an exception is thrown
     * @param pose initial pose of the new object
     * @param refFrame name of the world model object whose coordinate frame the pose is expressed in.  If nothing is provided here, the default is the robot "base" frame.
     */
    addFrame(name, pose, refFrame = 'base') {
        return this.addStatements(`add_frame("${name}", p[${pose.position[0]}, ${pose.position[1]}, ${pose.position[2]}, ${pose.orientation[0]}, ${pose.orientation[1]}, ${pose.orientation[2]}], "${refFrame}")`);
    }
    /**
     * Delete the frame named frame from the world model.
     * The "world", "base", and "tcp" frames cannot be deleted.
     * Any frames that are attached to the deleted frame will be attached to the "world" frame with new frame offsets set
     * such that the detached frame does not move in the world.
     * This command will fail if the frame does not exist.
     * @param name name of the frame to be deleted
     */
    deleteFrame(name) {
        return this.addStatements(`delete_frame("${name}")`);
    }
    /**
     * Changes the placement of the coordinate frame named name to the new placement given by pose that is defined in the refFrame coordinate frame.
     * This will fail if name is “world”, "tcp", or if the frame does not exist. Note: to move the "tcp" frame, use the set_tcp() command instead.
     * If being used with the part positioner product, the ref_name argument can be the name of an external axis or axis group.
     * @param name the name of the frame to move
     * @param pose the new placement
     * @param refFrame the coordinate frame that pose is expressed in.  The default value is the robot's "base" frame.
     */
    moveFrame(name, pose, refFrame = 'base') {
        return this.addStatements(`move_frame("${name}", p[${pose.position[0]}, ${pose.position[1]}, ${pose.position[2]}, ${pose.orientation[0]}, ${pose.orientation[1]}, ${pose.orientation[2]}], "${refFrame}")`);
    }
    /**
     * Attaches the child frame to the parent world model object. The relative transform between the parent and child will be set such that
     * the child does not move in the world when the attachment occurs.
     * The child cannot be "world", "tcp", or the same as parent.
     * This will fail if child or parent is not an existing frame, or this makes the attachments form a closed chain.
     * @param child name of the frame to be attached.  The name must not be "world" or "tcp".
     * @param parent name of the object that the child frame will be attached to.
     */
    attachFrame(child, parent) {
        return this.addStatements(`attach_frame("${child}", "${parent}")`);
    }
    /**
     * Changes the orientation of the frame @p name to the given @p orientation.
     */
    setFrameOrientation(name, orientation) {
        const baseFrameNameString = `"${name}"`;
        return this.addStatements([
            `local existingBaseParent = get_frame_parent(${baseFrameNameString})`,
            `local basePose = get_pose(${baseFrameNameString}, existingBaseParent)`,
            `basePose[3] = ${orientation[0]}`,
            `basePose[4] = ${orientation[1]}`,
            `basePose[5] = ${orientation[2]}`,
            `move_frame(${baseFrameNameString}, basePose, existingBaseParent)`,
        ].join('\n'));
    }
    /**
     * Add a popup with the given parameters.
     *
     * @param message    The popup message.
     * @param title      The popup title.
     * @param level      The popup level.
     * @param isBlocking true if its a blocking popup and false otherwise.
     * @return this object.
     */
    popup(message, title, level, isBlocking) {
        let popupScript;
        switch (level) {
            case PopupLevel.ERROR:
                popupScript = `popup("${message}", "${title}", False, True, blocking=${ScriptBuilder.getBooleanAsString(isBlocking)})`;
                break;
            case PopupLevel.WARNING:
                popupScript = `popup("${message}", "${title}", True, False, blocking=${ScriptBuilder.getBooleanAsString(isBlocking)})`;
                break;
            case PopupLevel.INFO:
                popupScript = `popup("${message}", "${title}", False, False, blocking=${ScriptBuilder.getBooleanAsString(isBlocking)})`;
                break;
        }
        this.addStatements(popupScript);
        return this;
    }
    /**
     * Request a value with the given parameters.
     *
     * @param variableName  The variable name to assign the value to
     * @param message       The request message.
     * @param valueType     The type: integer, float, boolean, or string
     * @return this object.
     */
    requestValue(variableName, message, valueType) {
        if (variableName && message && valueType !== null) {
            const func = `request_${valueType.toLowerCase()}_from_primary_client("${message}")`;
            this.globalVariable(variableName, func);
        }
        return this;
    }
    /**
     * Add a halt statement.
     *
     * @return this object.
     */
    halt() {
        this.addStatements('halt');
        return this;
    }
    /**
     * @deprecated Use {@link break}
     * @description Insert a break statement.
     * @return this object.
     */
    insertBreak() {
        return this.break();
    }
    /**
     * @description Insert a break statement.
     * @return this object.
     */
    break() {
        this.addStatements('break');
        return this;
    }
    /**
     * Add a sync command that uses up the remaining "physical" time a thread has in the current frame.
     *
     * @return this object.
     */
    sync() {
        this.addStatements('sync()');
        return this;
    }
    /**
     * Add a sleep command.
     *
     * @param seconds The amount of time in seconds.
     * @return this object.
     */
    sleep(seconds) {
        this.addStatements(`sleep(${seconds})`);
        return this;
    }
    /**
     * Add a stopl command. It decelerates tool speed to zero.
     *
     * @param toolAcceleration The tool acceleration in <i>m/</i><i>s</i><sup>2</sup>.
     * @return this object.
     */
    stopl(toolAcceleration) {
        this.addStatements(`stopl(${toolAcceleration})`);
        return this;
    }
    /**
     * Add a stopj command. It decelerates joint speeds to zero.
     *
     * @param jointAcceleration The joint acceleration in <i>rad/</i><i>s</i><sup>2</sup>.
     * @return this object.
     */
    stopj(jointAcceleration) {
        this.addStatements(`stopj(${jointAcceleration})`);
        return this;
    }
    /**
     * Add a call to the function that will accelerate linearly in Cartesian space and continue with constant tool
     * speed.
     *
     * @param xd                    array of tool speeds in m/s (spatial vector).
     * @param cartesianAcceleration tool position acceleration <i>m/</i><i>s</i><sup>2</sup>.
     * @param time                  time in seconds before the function returns.
     * @param refFrameName          name of reference frame
     * @return this object.
     */
    speedl(xd, cartesianAcceleration, time, refFrameName) {
        if (refFrameName) {
            this.addStatements(`speedl(struct(vel=${ScriptBuilder.formatArray(xd)}, frame="${refFrameName}"), ${cartesianAcceleration}, ${time})`);
        }
        else {
            this.addStatements(`speedl(${ScriptBuilder.formatArray(xd)}, ${cartesianAcceleration}, ${time})`);
        }
        return this;
    }
    /**
     * Add a call to the function that will accelerate linearly in joint space and continue with constant joint
     * speed.
     *
     * @param qd                array of joint speeds in rad/s.
     * @param jointAcceleration joint acceleration in <i>rad/</i><i>s</i><sup>2</sup> (of leading axis)
     * @param time              time in seconds before the function returns.
     * @return this object.
     */
    speedj(qd, jointAcceleration, time) {
        this.addStatements(`speedj(${ScriptBuilder.formatArray(qd)}, ${jointAcceleration}, ${time})`);
        return this;
    }
    /**
     * Add a call to the function that will Move to position (linear in joint-space).
     * @param q joint positions (q can also be specified as a pose, then
     * inverse kinematics is used to calculate the corresponding
     * joint positions) as a string or number array.
     * @param acc joint acceleration of leading axis in <i>rad/</i><i>s</i><sup>2</sup> as a string or number.
     * @param speed joint speed of leading axis in rad/s as a string or number (optional).
     * @param time time in seconds as a string or number (optional).
     * @param blendRadius blend radius in m as a string or number (optional).
     * @return this object.
     */
    movej(q, acc, speed, time, blendRadius) {
        if (typeof q !== 'string' && !(q instanceof String)) {
            q = ScriptBuilder.formatArray(q);
        }
        const args = [`${q}`];
        if (acc) {
            args.push(`a=${acc}`);
        }
        if (speed) {
            args.push(`v=${speed}`);
        }
        if (time) {
            args.push(`t=${time}`);
        }
        if (blendRadius) {
            args.push(`r=${blendRadius}`);
        }
        this.addStatements(`movej(${args.join(', ')})`);
        return this;
    }
    /**
     * Add a call to the function that will Move to position (linear in tool-space).
     * @param pose target pose (pose can also be specified as joint
     * positions, then forward kinematics is used to
     * calculate the corresponding pose) as a string.
     * @param acc tool acceleration <i>m/</i><i>s</i><sup>2</sup> as a string or number.
     * @param speed tool speed in m/s as a string or number (optional).
     * @param time time in seconds as a string or number (optional).
     * @param blendRadius blend radius in m as a string or number (optional).
     * @return this object.
     */
    movel(pose, acc, speed, time, blendRadius) {
        const args = [pose];
        if (acc) {
            args.push(`a=${acc}`);
        }
        if (speed) {
            args.push(`v=${speed}`);
        }
        if (time) {
            args.push(`t=${time}`);
        }
        if (blendRadius) {
            args.push(`r=${blendRadius}`);
        }
        this.addStatements(`movel(${args.join(', ')})`);
        return this;
    }
    /**
     * Blend circular (in tool-space) and move linear (in tool-space) to position. Accelerates to and moves
     * with constant tool speed v.
     * @param q joint positions (q can also be specified as a pose, then
     * inverse kinematics is used to calculate the corresponding
     * joint positions) as a string or number array.
     * @param acc joint acceleration of leading axis in <i>rad/</i><i>s</i><sup>2</sup> as a string or number.
     * @param speed joint speed of leading axis in rad/s as a string or number (optional).
     * @param blendRadius blend radius in m as a string or number (optional).
     * @return this object.
     */
    movep(q, acc, speed, blendRadius) {
        if (typeof q !== 'string' && !(q instanceof String)) {
            q = ScriptBuilder.formatArray(q);
        }
        const args = [`${q}`];
        if (acc) {
            args.push(`a=${acc}`);
        }
        if (speed) {
            args.push(`v=${speed}`);
        }
        if (blendRadius) {
            args.push(`r=${blendRadius}`);
        }
        this.addStatements(`movep(${args.join(', ')})`);
        return this;
    }
    /**
     * TCP moves on the circular arc segment from current pose, through pose_via to pose_to.
     * Accelerates to and moves with constant tool speed v. Use the mode parameter to define the
     * orientation interpolation.
     * @param pose_via path point (note: only position is used). Pose_via can also be specified as joint
     * positions, then forward kinematics is used to calculate the corresponding pose as a string or number array.
     * @param pose_to target pose (note: only position is used in Fixed orientation mode). Pose_to can also
     * be specified as joint positions, then forward kinematics is used to calculate the corresponding
     * pose as a string or number array.
     * @param acc tool acceleration <i>rad/</i><i>s</i><sup>2</sup> as a string or number (optional).
     * @param speed tool speed in m/s as a string or number (optional).
     * @param blendRadius blend radius (of target pose) in m as a string or number (optional).
     * @param mode mode of circular arc move as a string or number (optional).
     * 0: Unconstrained mode. Interpolate orientation from current pose to target pose (pose_to)
     * 1: Fixed mode. Keep orientation constant relative to the tangent of the circular arc (starting
     * from current pose)
     * @return this object.
     */
    movec(pose_via, pose_to, acc, speed, blendRadius, mode) {
        if (typeof pose_via !== 'string') {
            pose_via = ScriptBuilder.formatArray(pose_via);
        }
        if (typeof pose_to !== 'string') {
            pose_to = ScriptBuilder.formatArray(pose_to);
        }
        const args = [`${pose_via}`, `${pose_to}`];
        if (acc) {
            args.push(`a=${acc}`);
        }
        if (speed) {
            args.push(`v=${speed}`);
        }
        if (blendRadius) {
            args.push(`r=${blendRadius}`);
        }
        if (mode) {
            args.push(`mode=${mode}`);
        }
        this.addStatements(`movec(${args.join(', ')})`);
        return this;
    }
    /**
     * Add a call to the function that will Move to position (linear in joint-space), using OptiMove
     * @param q joint positions (q can also be specified as a pose, then
     * inverse kinematics is used to calculate the corresponding
     * joint positions) as a string or number array.
     * @param acc Joint acceleration as a fraction of what the joints are able to perform - a (0.0, 1.0] (optional).
     * @param speed Joint speed as a fraction of how fast the joints can move during the motion - v (0.0, 1.0] (optional).
     * @param blendRadius blend radius in m as a string or number (optional).
     * @return this object.
     */
    optimovej(q, acc, speed, blendRadius) {
        if (typeof q !== 'string' && !(q instanceof String)) {
            q = ScriptBuilder.formatArray(q);
        }
        const args = [`${q}`];
        if (acc) {
            args.push(`a=${acc}`);
        }
        if (speed) {
            args.push(`v=${speed}`);
        }
        if (blendRadius) {
            args.push(`r=${blendRadius}`);
        }
        this.addStatements(`optimovej(${args.join(', ')})`);
        return this;
    }
    /**
     * Add a call to the function that will Move to position (linear in tool-space), using OptiMove
     * @param pose target pose (pose can also be specified as joint
     * positions, then forward kinematics is used to
     * calculate the corresponding pose) as a string.
     * @param acc tool acceleration <i>m/</i><i>s</i><sup>2</sup> as a string or number.
     * @param speed tool speed in m/s as a string or number (optional).
     * @param blendRadius blend radius in m as a string or number (optional).
     * @return this object.
     */
    optimovel(pose, acc, speed, blendRadius) {
        const args = [pose];
        if (acc) {
            args.push(`a=${acc}`);
        }
        if (speed) {
            args.push(`v=${speed}`);
        }
        if (blendRadius) {
            args.push(`r=${blendRadius}`);
        }
        this.addStatements(`optimovel(${args.join(', ')})`);
        return this;
    }
    /**
     * @deprecated use {@link setTargetPayload}
     * @description Add a call to the function that sets the mass.
     * @param mass mass in kilograms as a Script.
     * @return this object.
     */
    setPayloadMass(mass) {
        this.addStatements(`set_payload_mass(${mass})`);
        return this;
    }
    /**
     * @deprecated use {@link setTargetPayload}
     * @description Add a call to the function that sets the mass and CoG.
     * @param mass mass in kilograms as a Script.
     * @param cx CoG x offset in meters
     * @param cy CoG y offset in meters
     * @param cz CoG z offset in meters
     * @param inertia The payload inertia matrix in kg*m^2, elements lxx,lyy
     * lzz, lxy, lxz, lyz
     * @return this object.
     */
    setPayloadMassAndCoG(mass, cx, cy, cz, inertia) {
        return this.setTargetPayload(mass, cx, cy, cz, inertia);
    }
    /**
     * @description Add a call to the function that sets the mass and CoG.
     * @param mass mass in kilograms as a Script.
     * @param cx CoG x offset in meters
     * @param cy CoG y offset in meters
     * @param cz CoG z offset in meters
     * @param inertia The payload inertia matrix in kg*m^2, elements lxx,lyy
     * lzz, lxy, lxz, lyz
     * @return this object.
     */
    setTargetPayload(mass, cx, cy, cz, inertia) {
        if (inertia) {
            this.addStatements(`set_target_payload(${mass}, [${cx}, ${cy}, ${cz}], [${inertia[0]}, ${inertia[1]}, ${inertia[2]}, ${inertia[3]}, ${inertia[4]}, ${inertia[5]}])`);
        }
        else {
            this.addStatements(`set_target_payload(${mass}, [${cx}, ${cy}, ${cz}])`);
        }
        return this;
    }
    /**
     * Set the direction of the acceleration experienced by the robot.
     *
     * @param x x m/s^2
     * @param y y m/s^2
     * @param z z m/s^2
     */
    setGravity(x, y, z) {
        return this.addStatements(`set_gravity([${x},${y},${z}])`);
    }
    /**
     * @description Add a call function which specifies the desired TCP
     * @param x x length
     * @param y y length
     * @param z z length
     * @param rx rx angle
     * @param ry ry angle
     * @param rz rz angle
     * @param name Name of the TCP
     */
    setTcp(x, y, z, rx, ry, rz, name) {
        this.addStatements(`set_tcp(p[${x}, ${y}, ${z}, ${rx}, ${ry}, ${rz}]${name ? ', "' + name + '"' : ''})`);
        return this;
    }
    /**
     * Append the content of another script builder to
     * this script builder
     *
     * @param scriptBuilder
     */
    append(scriptBuilder) {
        if (scriptBuilder.getIndention() < 0) {
            this.increaseIndent(scriptBuilder.getIndention());
            this.addStatements(scriptBuilder.getScript());
        }
        else {
            this.addStatements(scriptBuilder.getScript());
            this.increaseIndent(scriptBuilder.getIndention());
        }
        return this;
    }
    /**
     * The terminal operation that generates the as string containing the UR Script.
     *
     * @return the string that contains the UR Script.
     */
    getScript() {
        // @ts-ignore trimEnd is not part of the es2015 lib, even though it exists
        return this.script.trimEnd();
    }
    /**
     * Get the current indention level of this ScriptBuilder.
     *
     * @return the indention level as a number.
     */
    getIndention() {
        return this.currentIndent;
    }
    /**
     * Increase the indent.
     */
    increaseIndent(level = ScriptBuilder.SINGLE_INDENT_LEVEL) {
        this.currentIndent += level;
    }
    /**
     * Decrease the indent.
     */
    decreaseIndent(level = ScriptBuilder.SINGLE_INDENT_LEVEL) {
        this.currentIndent -= level;
    }
}

/**
 * Creates the serializer for the ScriptBuilder. Handles serializing and deserializing of the object.
 * @constructor
 */
const ScriptBuilderSerializer = () => {
    return {
        /**
         * Deserializes the message as a new ScriptBuilder based on the message inputs
         * @param message Message containing the values to be parsed to the new ScriptBuilder
         * @param defaultHandler The object to handle deserializing if not of type ScriptBuilder
         */
        deserialize(message, defaultHandler) {
            if (message && message.type === '$$ScriptBuilder') {
                const script = message.script;
                const currentIndent = message.currentIndent;
                return new ScriptBuilder(script, currentIndent);
            }
            return defaultHandler(message);
        },
        /**
         * Serializes the object and inserts a type to be used when deserializing to proper identifying the object when returning
         * @param item The original item that will be serialized
         * @param defaultHandler The object to handle serializing if not of type ScriptBuilder
         */
        serialize(item, defaultHandler) {
            if (item && item.constructor && item.constructor.name === 'ScriptBuilder') {
                return {
                    type: '$$ScriptBuilder',
                    ...item,
                };
            }
            return defaultHandler(item);
        },
    };
};

/**
 * Checks that the message coming is in a isSerializedScriptContext
 * @param message
 */
function isSerializedScriptContext(message) {
    return message && message.type === '$$ScriptContext';
}
/**
 * Creates the serializer for the ScriptContext. Handles serializing and deserializing of the object.
 * @constructor
 */
const ScriptContextSerializer = () => {
    return {
        /**
         * Deserializes the message as a new ScriptContext based on the message inputs
         * @param message Message containing the values to be parsed to the new ScriptContext
         * @param defaultHandler The object to handle deserializing if not of type ScriptContext
         */
        deserialize(message, defaultHandler) {
            if (isSerializedScriptContext(message)) {
                return {
                    traverse: {
                        ancestors: new AsyncArrayTraverser(message.ancestors ? message.ancestors : []),
                        children: new AsyncArrayTraverser(message.children ? message.children : []),
                        depthFirst: new AsyncArrayTraverser(message.depthFirst ? message.depthFirst : []),
                        nextSiblings: new AsyncArrayTraverser(message.nextSiblings ? message.nextSiblings : []),
                        previousSiblings: new AsyncArrayTraverser(message.previousSiblings ? message.previousSiblings : []),
                    },
                };
            }
            else {
                return defaultHandler(message);
            }
        },
        /**
         * Serializes the object and inserts a type to be used when deserializing to proper identifying the object when returning
         * @param item The original item that will be serialized
         * @param defaultHandler The object to handle serializing if not of type ScriptContext
         */
        serialize(item, defaultHandler) {
            if (isSerializedScriptContext(item)) {
                return {
                    type: '$$ScriptContext',
                    ancestors: item.ancestors,
                    children: item.children,
                    depthFirst: item.depthFirst,
                    previousSiblings: item.previousSiblings,
                    nextSiblings: item.nextSiblings,
                };
            }
            else {
                return defaultHandler(item);
            }
        },
    };
};

/**
 * Checks that the message coming is in a isSerializedValidationContext
 * @param message
 */
function isSerializedValidationContext(message) {
    return message && message.type === '$$ValidationContext';
}
/**
 * Creates the serializer for the ValidationContext. Handles serializing and deserializing of the object.
 * @constructor
 */
const ValidationContextSerializer = () => {
    return {
        /**
         * Deserializes the message as a new ValidationContext based on the message inputs
         * @param message Message containing the values to be parsed to the new ValidationContext
         * @param defaultHandler The object to handle deserializing if not of type ValidationContext
         */
        deserialize(message, defaultHandler) {
            if (isSerializedValidationContext(message)) {
                return {
                    traverse: {
                        ancestors: new AsyncArrayTraverser(message.ancestors ? message.ancestors : []),
                        children: new AsyncArrayTraverser(message.children ? message.children : []),
                        depthFirst: new AsyncArrayTraverser(message.depthFirst ? message.depthFirst : []),
                        nextSiblings: new AsyncArrayTraverser(message.nextSiblings ? message.nextSiblings : []),
                        previousSiblings: new AsyncArrayTraverser(message.previousSiblings ? message.previousSiblings : []),
                    },
                };
            }
            else {
                return defaultHandler(message);
            }
        },
        /**
         * Serializes the object and inserts a type to be used when deserializing to proper identifying the object when returning
         * @param item The original item that will be serialized
         * @param defaultHandler The object to handle serializing if not of type ValidationContext
         */
        serialize(item, defaultHandler) {
            if (isSerializedValidationContext(item)) {
                return {
                    type: '$$ValidationContext',
                    ancestors: item.ancestors,
                    children: item.children,
                    depthFirst: item.depthFirst,
                    previousSiblings: item.previousSiblings,
                    nextSiblings: item.nextSiblings,
                };
            }
            else {
                return defaultHandler(item);
            }
        },
    };
};

/**
 * Setup registration of serializers to be used when sending/receiving object across worker boundaries
 */
function registerSerializers() {
    registerSerializer(ScriptContextSerializer());
    registerSerializer(ValidationContextSerializer());
    registerSerializer(InsertionContextSerializer());
    registerSerializer(ScriptBuilderSerializer());
    registerSerializer(ApplicationContextSerializer());
}

/**
 * Common handler for behaviors
 */
const behaviorHandler = (behaviors) => {
    registerSerializers();
    expose(behaviors);
};
/**
 * Registers a Program behavior object allowing the programming area to
 * call the behavior methods.
 *
 * @param behaviors The ProgramBehaviors object to register
 */
const registerProgramBehavior = behaviorHandler;
/**
 * Registers an Application behavior object allowing the application area to
 * call the behavior methods.
 *
 * @param behaviors The ApplicationBehaviors object to register
 */
const registerApplicationBehavior = behaviorHandler;
/**
 * Registers a Source behavior object allowing various entities to call source
 * behavior methods
 *
 * @param behaviors The SourceBehaviors object to register
 */
const registerSourceBehavior = behaviorHandler;
/**
 * Registers a Smart skill behavior object allowing each smart skill to call its execution methods
 * behavior methods
 *
 * @param behaviors The SmartSkillsBehaviors object to register
 */
const registerSmartSkillBehavior = behaviorHandler;
/**
 * Registers a Operator Screen behavior object allowing each operator screen to call its
 * behavior methods
 *
 * @param behaviors The OperatorScreenBehaviors object to register
 */
const registerOperatorScreenBehavior = behaviorHandler;
/**
 * Registers a System Info behavior object allowing each system info screen to call its
 * behavior methods
 *
 * @param behaviors The SystemInfoBehaviors object to register
 */
const registerSystemInfoBehavior = behaviorHandler;
/**
 * Registers a Sidebar behavior object allowing each sidebar to call its
 * behavior methods
 *
 * @param behaviors The SidebarBehaviors object to register
 */
const registerSidebarBehavior = behaviorHandler;

const VALIDATION_OK = { isValid: true };
const VALIDATION_INVALID = { isValid: false };

/**
 * API for operator screen presenter components
 */
class OperatorScreenPresenterAPI extends CommonPresenterAPI {
    /** Creates a new instance of the OperatorScreenPresenterAPI, to allow communication with the
     * surrounding application from within a WebComponent
     *
     * @param target An EventTarget, provided by PolyScope X
     */
    constructor(target) {
        super(target);
        this.robotMoveService = new RobotMoveService(target);
        this.programExecutionService = new ProgramExecutionService(target);
        this.toolService = new ToolService(target);
        this.mountingService = new MountingService(target);
        this.framesService = new FramesService(target);
        this.operatorScreenService = new OperatorScreenService(target);
    }
}
/**
 * Provides access to Operator Screen related API features in BehaviorWorkers
 */
class OperatorScreenBehaviorAPI extends CommonBehaviorAPI {
    /** Creates a new instance of the OperatorScreenBehaviorAPI to allow communication with the
     * surrounding application from within a WebWorker
     *
     * @param target A worker, provided with 'self' when created inside a web worker
     */
    constructor(target) {
        super(target);
    }
}

/**
 * WebComponentDialogAPI
 */
class WebComponentDialogAPI extends CommonPresenterAPI {
    /** Creates a new instance of the WebComponentDialogAPI, to allow communication with the
     * surrounding application from within a WebComponent
     *
     * @param target An EventTarget, provided by PolyScope X
     */
    constructor(target) {
        super(target);
        this.robotMoveService = new RobotMoveService(target);
    }
}

/**
 * Provides access to system info related API features in System Info Presenters
 */
class SystemInfoPresenterAPI extends CommonPresenterAPI {
    /** Creates a new instance of the SystemInfoPresenterAPI, to allow communication with the
     * surrounding application from within a WebComponent
     *
     * @param target An EventTarget, provided by PolyScope X
     */
    constructor(target) {
        super(target);
    }
}

/**
 * Provides access to Sidebar related API features in BehaviorWorkers
 */
class SidebarBehaviorAPI extends CommonBehaviorAPI {
    /**
     * Creates a new instance of the SidebarBehaviorAPI to allow communication with the
     * surrounding application from within a WebWorker
     *
     * @param target A worker, provided with 'self' when created inside a web worker
     */
    constructor(target) {
        super(target);
    }
}
/**
 * Provides access to Sidebar related API features in Presenters
 */
class SidebarPresenterAPI extends CommonPresenterAPI {
    /**
     * Create a new instance of the SidebarPresenterAPI to allow a Presenter to interact with the API and update itself
     */
    constructor(target) {
        super(target);
    }
}

/**
 * Model to be used in collaboration with tabInputComponent.
 */
class TabInputModel {
    /**
     * Initialize the object.
     */
    constructor(entity, selectedType, value) {
        this.entity = entity;
        this.selectedType = selectedType;
        this.value = value;
    }
    /**
     * Validates the model
     * @param model to validate
     */
    static async isValid(model) {
        if (model.selectedType === SelectedInput.VARIABLE) {
            const api = new ProgramBehaviorAPI(self);
            const isSuppressed = await api.symbolService.isSuppressed(model.value);
            const isRegisteredVariableName = await api.symbolService.isRegisteredVariableName(model.value);
            if (!isRegisteredVariableName || isSuppressed) {
                return false;
            }
        }
        return true;
    }
    /**
     * Checks if tabInputModel has an entity and a selectedType
     * @param value to be interrogated
     */
    static isTabInputModel(value) {
        return value.entity !== undefined && value.selectedType !== undefined;
    }
    /**
     * Returns TabInputValue from TabInputModel which is the input for the TabInput component
     */
    static getTabInputValue(model) {
        return {
            selectedType: model.selectedType,
            value: model.value,
        };
    }
    /**
     * Sets TabInputValue on TabInputModel which is the input for the TabInput component
     * @param model to update the TabInputValue
     * @param value the inserted value for the model.
     */
    static setTabInputValue(model, value) {
        model.selectedType = value.selectedType;
        model.value = value.value;
    }
}

/**
 * Describes a relative insertion point
 */
var InsertionEnum;
(function (InsertionEnum) {
    InsertionEnum["BEFORE"] = "before";
    InsertionEnum["AFTER"] = "after";
    InsertionEnum["INTO_FIRST"] = "intoFirst";
    InsertionEnum["INTO_LAST"] = "intoLast";
})(InsertionEnum || (InsertionEnum = {}));

const AngleUnits = ['rad', 'deg'];

const CurrentUnits = ['A', 'mA'];

const AccelerationUnits = ['m/s^2', 'mm/s^2'];

const AngularAccelerationUnits = ['rad/s^2', 'deg/s^2'];

const AngularSpeedUnits = ['rad/s', 'deg/s'];

const ForceUnits = ['N'];

const LengthUnits = ['m', 'mm'];

const MassUnits = ['g', 'kg'];

const SpeedUnits = ['m/s', 'mm/s'];

const TimeUnits = ['s', 'ms'];

const MomentInertiaUnits = ['kg*m^2'];

const MomentumUnits = ['kg⋅m/s'];

const TemperatureUnits = ['C'];

const VoltageUnits = ['V', 'mV'];

const AllUnitStrings = new Set([
    ...MassUnits,
    ...SpeedUnits,
    ...LengthUnits,
    ...AngleUnits,
    ...VoltageUnits,
    ...CurrentUnits,
    ...AngularSpeedUnits,
    ...AccelerationUnits,
    ...AngularAccelerationUnits,
    ...TimeUnits,
    ...ForceUnits,
    ...MomentumUnits,
    ...MomentInertiaUnits,
    ...TemperatureUnits,
]);

/**
 * Map object containing all the ValueObject strings, mapped to their Unit Library unit
 */
const ValueObjectMap = {
    g,
    kg,
    kgM,
    'kg·m/s': kgMPerSec,
    'deg/s': degPerSec,
    'deg/s^2': degPerSecSq,
    'm/s': mPerSec,
    'm/s^2': mPerSecSq,
    'mm/s': mmPerSec,
    'mm/s^2': mmPerSecSq,
    'rad/s': radPerSec,
    'rad/s^2': radPerSecSq,
    A,
    N,
    V,
    deg,
    m,
    mA,
    mV,
    mm,
    ms,
    rad,
    s,
    W,
    'kg*m^2': kgMSq,
    degreesCelsius,
};

/**
 * Converts one value object to another one with a different compatible unit
 * @example
 * ```ts
 *  const value = { value: 1, unit: m/s };
 *  const toUnit = mm/s;
 *  const convertedValue = convertValue(value, toUnit);
 *  console.log(convertedValue); // outputs { value: 1000, unit: mm/s };
 * ```
 * @throws Error if the value object and the provided unit are incompatible
 * @param value The value object to be converted
 * @param toUnit The unit to convert into, must be compatible with the value object's unit
 * @returns a new value object that contains the converted value
 * @see valueRawConverter
 */
function convertValue(value, toUnit) {
    if (!Object.keys(ValueObjectMap).includes(value.unit)) {
        throw new Error(`Unknown unit in value ${value.unit}`);
    }
    if (!Object.keys(ValueObjectMap).includes(toUnit)) {
        throw new Error(`Unknown target unit ${toUnit}`);
    }
    const from = ValueObjectMap[value.unit];
    const to = ValueObjectMap[toUnit];
    if (from.quantity.quantity !== to.quantity.quantity) {
        throw new Error('Cannot convert mismatched units');
    }
    const originalValue = new FixedPointNumber(value.value, from);
    const newValue = converter(originalValue, to);
    return {
        value: newValue.getValue(),
        unit: toUnit,
    };
}
/**
 * Get the raw numeric value of a value object converted into another compatible unit as a string
 * <p>The returned value is formatted using {@link FixedPointNumber#getDisplayValue} and is without any unit symbols or label</p>
 * @example
 * ```ts
 *  const value = { value: 1, unit: "m/s" };
 *  const toUnit = "mm/s";
 *  const convertedValue = valueRawConverter(value, toUnit);
 *  console.log(convertedValue); // outputs "1000"
 * ```
 * @throws Error if the value object and the provided unit are incompatible
 * @param value The value object to be converted
 * @param to The unit to convert into, must be compatible with the value object's unit
 * @returns The raw numeric value of the converted value object as a formatted string
 * @see convertValue
 */
function valueRawConverter(value, to) {
    const convertedValue = convertValue(value, to.label);
    return `${new FixedPointNumber(Number(convertedValue.value), to).getDisplayValue()}`;
}
/**
 * Transform the value object into a formatted string converting it to the specified unit with the unit's symbol appended
 * <p>The returned value is formatted using {@link FixedPointNumber#getDisplayValue} and is without any unit symbols or label</p>
 * @example
 * ```ts
 *  const value = { value: 1, unit: "m/s" };
 *  const toUnit = "mm/s";
 *  const convertedValue = valueConverter(value, toUnit);
 *  console.log(convertedValue); // outputs "1000 mm/s"
 * ```
 * @throws Error if the value object and the provided unit are incompatible
 * @param value The value object to be formatted
 * @param to The unit to convert into, must be compatible with the value object's unit
 * @returns The formatted value with the unit's symbol appended
 */
function valueConverter(value, to) {
    const convertedValue = convertValue(value, to.label);
    return `${new FixedPointNumber(Number(convertedValue.value), to).getDisplayValue()} ${to.symbol}`;
}

/**
 * Presets available for Digital In signals
 */
var DigitalInPreset;
(function (DigitalInPreset) {
    DigitalInPreset[DigitalInPreset["NONE"] = 0] = "NONE";
    DigitalInPreset[DigitalInPreset["START"] = 1] = "START";
    DigitalInPreset[DigitalInPreset["STOP"] = 2] = "STOP";
    DigitalInPreset[DigitalInPreset["PAUSE"] = 3] = "PAUSE";
})(DigitalInPreset || (DigitalInPreset = {}));

/**
 * Enumeration for possible signal directions
 */
var SignalDirectionEnum;
(function (SignalDirectionEnum) {
    SignalDirectionEnum["IN"] = "IN";
    SignalDirectionEnum["OUT"] = "OUT";
})(SignalDirectionEnum || (SignalDirectionEnum = {}));
/**
 * Enumeration for possible signal value types
 */
var SignalValueTypeEnum;
(function (SignalValueTypeEnum) {
    SignalValueTypeEnum["BOOLEAN"] = "BOOLEAN";
    SignalValueTypeEnum["FLOAT"] = "FLOAT";
    SignalValueTypeEnum["REGISTER"] = "REGISTER";
    SignalValueTypeEnum["CUSTOM"] = "CUSTOM";
})(SignalValueTypeEnum || (SignalValueTypeEnum = {}));

/**
 * Enum with all possible analog domain values
 */
var SignalAnalogDomainValueEnum;
(function (SignalAnalogDomainValueEnum) {
    SignalAnalogDomainValueEnum["CURRENT"] = "CURRENT";
    SignalAnalogDomainValueEnum["VOLTAGE"] = "VOLTAGE";
})(SignalAnalogDomainValueEnum || (SignalAnalogDomainValueEnum = {}));

/*
 * Public API Surface of contribution-api
 */

/**
 * Generated bundle index. Do not edit.
 */

export { APPLICATION_NODE_SERVICE_MESSAGE_TYPE_BASE, APPLICATION_SERVICE_MESSAGE_TYPE_BASE, AccelerationUnits, AllUnitStrings, AngleUnits, AngularAccelerationUnits, AngularSpeedUnits, ApplicationBehaviorAPI, ApplicationNodeService, ApplicationNodeType, ApplicationPresenterAPI, ApplicationService, AsyncArrayTraverser, CenterOfGravityAxes, CurrentUnits, DIALOG_SERVICE_MESSAGE_TYPE_BASE, DialogService, DigitalInPreset, DurabilityPolicy, FRAMES_SERVICE_MESSAGE_TYPE_BASE, FREEDRIVE_STATE_SERVICE_MESSAGE_TYPE_BASE, ForceUnits, FramesService, FreedriveService, HistoryPolicy, InertiaMatrixKeys, InsertionEnum, LOG_MESSAGE_SERVICE_MESSAGE_TYPE_BASE, LOG_PERFORMANCE_SERVICE_MESSAGE_TYPE_BASE, LengthUnits, LivelinessPolicy, LogMessageService, LogPerformanceService, MOTION_PROFILES_SERVICE_MESSAGE_TYPE_BASE, MOUNTING_SERVICE_MESSAGE_TYPE_BASE, MassUnits, MessageInvoker, MessageInvokerService, MomentInertiaUnits, MomentumUnits, MotionProfileMoveType, MotionProfilesService, MountingService, MovementType, NodeType, OPERATOR_SCREEN_SERVICE_MESSAGE_TYPE_BASE, OperatorInputType, OperatorScreenBehaviorAPI, OperatorScreenPresenterAPI, OperatorScreenService, OrientationAxes, PROGRAM_CODE_SERVICE_MESSAGE_TYPE_BASE, PROGRAM_EXECUTION_SERVICE_MESSAGE_TYPE_BASE, PROGRAM_NODE_SERVICE_MESSAGE_TYPE_BASE, PROGRAM_SERVICE_MESSAGE_TYPE_BASE, PROGRAM_TREE_SERVICE_MESSAGE_TYPE_BASE, PopupLevel, PoseAxes, PositionAxes, ProgramBehaviorAPI, ProgramCodeService, ProgramExecutionService, ProgramNodeService, ProgramPresenterAPI, ProgramService, ProgramTreeService, QoSProfile, ROBOT_INFO_SERVICE_MESSAGE_TYPE_BASE, ROBOT_MOVE_SERVICE_MESSAGE_TYPE_BASE, ROBOT_POSITION_SERVICE_MESSAGE_TYPE_BASE, ReliabilityPolicy, RobotInfoService, RobotMoveService, RobotPositionService, Ros2Client, Ros2EventService, SAFETY_SERVICE_MESSAGE_TYPE_BASE, SETTINGS_MESSAGE_TYPE_BASE, SMART_SKILL_INSTANCE_SERVICE_MESSAGE_TYPE_BASE, SNACKBAR_SERVICE_MESSAGE_TYPE_BASE, SOURCE_MESSAGE_TYPE_BASE, SOURCE_NODE_SERVICE_MESSAGE_TYPE_BASE, SYMBOL_SERVICE_MESSAGE_TYPE_BASE, SYSTEM_SERVICE_MESSAGE_TYPE_BASE, SafetyService, ScriptBuilder, SettingsService, SidebarBehaviorAPI, SidebarPresenterAPI, SignalAnalogDomainValueEnum, SignalDirectionEnum, SignalService, SignalValueTypeEnum, SmartSkillInstanceService, SmartSkillType, SmartSkillsBehaviorAPI, SmartSkillsPresenterAPI, SnackbarService, SourceBehaviorAPI, SourceNodeService, SourcePresenterAPI, SourceService, SpeedUnits, SymbolService, SystemInfoPresenterAPI, SystemService, TOOL_SERVICE_MESSAGE_TYPE_BASE, TREE_BUILDER_MESSAGE_TYPE_BASE, TabInputModel, TemperatureUnits, ThreadAction, TimeUnits, ToolService, TreeBuilder, UNSUBSCRIBE_MESSAGE_TYPE, UREventTarget, URFunction, URLogicProgram, URModule, URSymbol, URSymbolType, URVariable, VALIDATION_INVALID, VALIDATION_OK, VALIDATION_SERVICE_MESSAGE_TYPE_BASE, ValidationService, ValueObjectMap, VariableValueType, VoltageUnits, WaypointType, WebComponentDialogAPI, convertValue, floatOperators, isFunction, isJointPositions, isLogicProgram, isModule, isReferenceItem, isURSymbol, isValueItem, isVariable, isWaypoint, iskinematicPosition, operatorInverseMap, postAndAwait, postAndObserve, registerApplicationBehavior, registerOperatorScreenBehavior, registerOperators, registerProgramBehavior, registerSerializers, registerSidebarBehavior, registerSmartSkillBehavior, registerSourceBehavior, registerSystemInfoBehavior, valueConverter, valueRawConverter };
