// @universal-robots/contribution-api/fesm2022/universal-robots-contribution-api-testing.mjs
/**
 * Stub for an async iterable iterator.
 */
function getAsyncIterableIteratorStub() {
    return {
        /**
         * @returns {AsyncIterableIterator<Readonly<ProgramNode>>} An async iterator.
         */
        [Symbol.asyncIterator]() {
            throw new Error('Method not implemented. This is only used as a stub for serialization.');
        },
        /**
         * @returns {Promise<IteratorResult<Readonly<ProgramNode>>>} The next item in the iterator.
         */
        next() {
            throw new Error('Method not implemented. This is only used as a stub for serialization.');
        },
    };
}

/**
 * Testing utility for serialization of a ScriptContext.
 */
class ScriptContextForWorkerSerialization {
    /**
     * Initializes a new instance of the ScriptContextForWorkerSerialization class.
     */
    constructor() {
        this.type = '$$ScriptContext';
        this.traverse = {
            ancestors: getAsyncIterableIteratorStub(),
            children: getAsyncIterableIteratorStub(),
            depthFirst: getAsyncIterableIteratorStub(),
            nextSiblings: getAsyncIterableIteratorStub(),
            previousSiblings: getAsyncIterableIteratorStub(),
        };
    }
}

/**
 * Testing utility for serialization of an InsertionContext.
 */
class InsertionContextForWorkerSerialization {
    /**
     * Initializes a new instance of the InsertionContextForWorkerSerialization class.
     */
    constructor() {
        this.type = '$$ScriptContext';
        this.traverse = {
            ancestors: getAsyncIterableIteratorStub(),
            nextSiblings: getAsyncIterableIteratorStub(),
            previousSiblings: getAsyncIterableIteratorStub(),
        };
    }
}

/**
 * Generated bundle index. Do not edit.
 */

export { InsertionContextForWorkerSerialization, ScriptContextForWorkerSerialization, getAsyncIterableIteratorStub };
